
---
title: "Testing issues with ALEs in data set"
author: "Mackenzie Johnson"
date: "10/27/2022"
output: html_document
---



```{r, message=FALSE}
library(tidyverse)
library(here)
library(glue)
library(reshape2)
library(ggtext)
library(broom)
```

## Read in the data
```{r}

# 48 out of 103
meta_df <- read_tsv(here("Data", "processed_data", "processed_metadata_ecoli.tsv")) %>% 
  filter(`Evolved Sample` == "No")

conditions <- meta_df$Simple_sample_id

exp_df <- read_tsv(
  here(
    "Data", 
    "processed_data", 
    "processed_expression_ecoli.tsv"
  )
) %>% 
  select_if(
    names(.) == "log-TPM" | 
      names(.) %in% conditions
  ) 

# # need to recalculate
# exp_sum_df <- read_tsv(
#   here(
#     "Data", 
#     "processed_data",
#     "processed_expression_summary_ecoli.tsv"
#   )
# ) %>% 
#   select(`log-TPM`) %>% View()

full_df <- read_tsv(here("Data", "ecoli_info", "current_ecoli_master_table.tsv")) %>% 
  drop_na(CAI)

```

# 4 explore expression
```{r}

cond_cor_mat <- exp_df %>% 
  select(-`log-TPM`) %>% 
  cor(., method = "pearson")

fig_exp_cond_corr <- melt(cond_cor_mat^2) %>% 
  ggplot(aes(x = Var1, y = Var2, fill = value)) +
  geom_tile() +
  coord_fixed() +
  scale_fill_viridis_c(
    option = "B", name = "*R*^2"
  ) +
  theme_void() +
  theme(legend.title = element_markdown())

```

```{r}
# find max pearson's r value between two conditions
test_mat <- cond_cor_mat
diag(test_mat) <- 0 # exclude self correlations
inds <- which(
  test_mat == max(test_mat), 
  arr.ind = TRUE
)

# store data, condition names, and replace column names for use across datasets
max_cond_corr_df <- tibble(
  exp_df[, inds[1]+1],
  exp_df[, inds[2]+1]
)
max_names <- colnames(max_cond_corr_df)
colnames(max_cond_corr_df) <- c("cond1", "cond2")

x_pos <- 3
y_pos <- 13.5
xy_min <- 0
xy_max <- 15

# create dataframe with R^2 values to plot
r2_max_text <- data.frame(
  label = glue(
    '*R*<sup>2</sup> = {round(cond_cor_mat[inds[1], inds[2]]*
    cond_cor_mat[inds[1], inds[2]], 3)}'),
  x = x_pos, 
  y = y_pos
)

# visualize two conditions with largest correlation
fig_max <- max_cond_corr_df %>% 
  ggplot(aes(x = cond1, y = cond2)) +
  geom_point(color = "black", alpha = 0.3) +
  #coord_fixed() +
  scale_x_continuous(
    limits = c(xy_min, xy_max), 
    expand = c(0.01, 0.01),
    name = glue('Condition {inds[1]} expression')
    # name = glue('Condition {inds[1]}: {max_names[1]}')
  ) +
  scale_y_continuous(
    limits = c(xy_min, xy_max),
    expand = c(0.02, 0.02),
    name = glue('Condition {inds[2]} expression')
  ) +
  geom_richtext(
    data = r2_max_text,
    aes(x = x, y = y,label = label),
    label.color = NA,
    inherit.aes = FALSE
  ) +
  theme_half_open() +
  background_grid()

max_names

# find min pearson's r value between two conditions
inds_min <- which(cond_cor_mat == min(cond_cor_mat), arr.ind = TRUE)

# store data, condition names, and replace column names for use across datasets
min_cond_corr_df <- tibble(
  exp_df[, inds_min[1]+1],
  exp_df[, inds_min[2]+1]
)
min_names <- colnames(min_cond_corr_df)
colnames(min_cond_corr_df) <- c("cond1", "cond2") # rename for easier reproducibility

# store R^2 value for use in plot
r2_min_text <- data.frame(
  label = glue(
    '*R*<sup>2</sup> = {round(cond_cor_mat[inds_min[1], inds_min[2]]*
    cond_cor_mat[inds_min[1], inds_min[2]], 3)}'),
  x = x_pos, 
  y = y_pos
)

# visualize the conditions with the smallest correlation
fig_min <- min_cond_corr_df %>% 
  ggplot(aes(x = cond1, y = cond2)) +
  geom_point(color = "black", alpha = 0.3) +
  #coord_fixed() +
  scale_x_continuous(
    limits = c(xy_min, xy_max),
    expand = c(0.01, 0.01),
    name = glue('Condition {inds_min[1]} expression')
  ) +
  scale_y_continuous(
    limits = c(xy_min, xy_max),
    expand = c(0.02, 0.02),
    name = glue('Condition {inds_min[2]} expression')
  ) +
  geom_richtext(
    data = r2_min_text,
    aes(x = x, y = y,label = label),
    label.color = NA,
    inherit.aes = FALSE
  ) +
  theme_half_open() +
  background_grid()

min_names

fig_2b <- plot_grid(fig_max, fig_min, nrow = 2)

fig_2 <- plot_grid(
  fig_exp_cond_corr, fig_2b,
  nrow = 1,
  labels = "AUTO",
  scale = c(1, 0.85)
)

fig_2

rm(fig_2b, fig_exp_cond_corr, fig_min, fig_max)
rm(test_mat, r2_max_text, r2_min_text, max_cond_corr_df, min_cond_corr_df)
rm(x_pos, y_pos, xy_max, xy_min)

```

```{r}

exp_gr_df <- exp_df %>% 
  gather(
    -`log-TPM`,
    key = "condition",
    value = "expression"
  ) %>% 
  spread(`log-TPM`, expression)

exp_gr_df <- meta_df %>% 
    select(`Simple_sample_id`, `Growth Rate (1/hr)`) %>% 
    inner_join(., exp_gr_df, by = c("Simple_sample_id" = "condition"))

exp_gr_corr <- cor(exp_gr_df[2], exp_gr_df[3:ncol(exp_gr_df)], method = "pearson") %>% 
   as_tibble() %>% 
   gather(
     .,
     key = "genes",
     value = "pearsons"
    ) %>%
   mutate(r2 = pearsons^2)
  
perm_mat <- lapply(exp_gr_df[3:ncol(exp_gr_df)], sample) %>% as_tibble()
exp_gr_perm <- cor(exp_gr_df[2], perm_mat, method = "pearson") %>% 
  as_tibble() %>% 
  gather(
    .,
    key = "genes",
    value = "pearsons"
  ) %>% 
  mutate(r2 = pearsons^2)
  
perm_df <- tibble(
  exp_gr_corr$pearsons,
  exp_gr_perm$pearsons
) %>%
  gather(., "set", "pearsons")

exp_gr_corr %>% filter(pearsons == max(pearsons)) %>% .$genes
exp_gr_corr %>% filter(pearsons == min(pearsons)) %>% .$genes

make_cor_plot <- function(x, gname) {
  x <- enquo(x)
  exp_gr_df %>%
    ggplot() +
    aes(!!x, `Growth Rate (1/hr)`) +
    geom_point(alpha = 0.75) +
    scale_x_continuous("Expression across conditions") +
    theme_minimal_grid() +
    ggtitle(glue("{gname}"))
}

fig_pos_cor <- make_cor_plot(b0154, "hemL")
fig_ant_cor <- make_cor_plot(b2106, "rcnA")

fig_perm <- perm_df %>% 
  ggplot() +
    aes(x = pearsons, fill = set) +
    geom_density(alpha = 0.5) +
    scale_x_continuous(
      name = "Pearson's *r* (individual gene expression vs growth)<br />**GCI**"
    ) +
    scale_y_continuous(name = "Density") +
    scale_fill_manual(
      name = "",
      labels = c("True data", "Shuffled data"),
      values = c("#009E73", "#808080")
    ) +
    theme_minimal_grid() +
    theme(
      legend.position = "top",
      axis.title.x = element_markdown()
    )

fig_4 <- (fig_ant_cor | fig_pos_cor) / fig_perm +
   plot_annotation(
     tag_levels = "A"
   )

fig_4

# range of GCI values -0.726333  0.710021

set1 <- perm_df %>% 
  filter(set == "exp_gr_corr$pearsons") %>% 
  select(pearsons)

set2 <- perm_df %>% 
  filter(set == "exp_gr_perm$pearsons") %>% 
  select(pearsons)

t.test(set1, set2)  # p-value < 2.2e-16


```

```{r}
fig_2_fixed <- cowplot::ggdraw(fig_2) + 
  theme(plot.background = element_rect(fill="white", color = NA))

save_plot(
  here(
    "Manuscript",
    #"Figures",
    glue('fig_corr_no_evol.png')   # file name
  ),
  fig_2_fixed, 
  ncol = 1, 
  nrow = 1, 
  base_height = 5.21, #base_height = 5.71,
  base_asp = 1.618, 
  base_width = NULL
)

save_plot(
  here(
    "Manuscript",  
    #"Figures",
    glue('fig_gene_con_no_evol.png')   # file name
  ),
  fig_4, 
  ncol = 1, 
  nrow = 1, 
  base_height = 6.71, #base_height = 5.71,
  base_asp = 1.618, 
  base_width = NULL
)

rm(exp_gr_corr, exp_gr_df, exp_gr_perm, fig_2, fig_2_fixed, fig_4, fig_ant_cor, fig_perm, fig_pos_cor, inds, inds_min, perm_df, perm_mat, set1, set2, make_cor_plot, cond_cor_mat, min_names, max_names)

```


# 5 explore cub
Need to check this:
```{r}

exp_sum_df <- exp_df %>% 
  left_join(., full_df, by = c("log-TPM" = "locus_tag")) %>% 
  rowwise() %>% 
  mutate(
    mean = mean(c_across(`fur__wt_fe`:`ytf__delyheO`))
  ) %>%
  select(`log-TPM`, mean) 

exp_sum_df <- exp_sum_df %>% 
  left_join(., exp_gr_corr, by = c("log-TPM" = "genes")) %>% 
  select(-r2) 

colnames(exp_sum_df) <- c("log-TPM", "mean", "lin_r")

# grow_rates <- meta_df %>% 
#   select(Simple_sample_id, `Growth Rate (1/hr)`)

# exp_sum_df <- exp_df %>% 
#   gather(
#     "condition", 
#     "expression", 
#     `fur__wt_fe`:`ytf__delyheO`
#   ) %>% 
#   left_join(
#     ., 
#     grow_rates, 
#     by = c("condition" = "Simple_sample_id")
#   ) %>% 
#   nest(data = -`log-TPM`) %>% 
#   mutate(
#     lin_r = map(
#       data, 
#       ~cor(.$`Growth Rate (1/hr)`, .$expression, method = "pearson")
#     )
#   ) %>% 
#   unnest(cols = lin_r) %>% 
#   select(`log-TPM`, lin_r) %>% 
#   left_join(., exp_sum_df)

```

```{r}
full_df <- inner_join(
  full_df, 
  exp_sum_df, 
  by = c("locus_tag" = "log-TPM")
)

rm(exp_sum_df)

# get df
corr_df <- full_df %>% 
  select(locus_tag, CAI) %>% 
  left_join(., exp_df, by = c("locus_tag" = "log-TPM")) 
# correlations
exp_cub_corr <- cor(corr_df$CAI, corr_df[3:50], method = "pearson") %>% 
  as_tibble() %>% 
  gather(
    .,
    key = "condition",
    value = "pearsons"
  ) %>% 
  mutate(r2 = pearsons^2)

min_corr <- exp_cub_corr %>% 
  filter(r2 == min(r2))
max_corr <- exp_cub_corr %>% 
  filter(r2 == max(r2))

fig_corr_min <- corr_df %>% 
  ggplot() +
  aes(x = `ica__cytd_rib`, y = CAI) +
  geom_point(alpha = 0.5) +
  scale_x_continuous(name = "Condition expression (log TPM)") +
  theme_minimal_grid() +
  labs(title = glue('*R*<sup>2</sup> = {round(min_corr$r2, 3)}')) +
  theme(plot.title = element_markdown())

fig_corr_max <- corr_df %>%
  ggplot() +
  aes(x = `rpoB__rpoBE672K_lb`, y = CAI) +
  geom_point(alpha = 0.5) +
  scale_x_continuous(name = "Condition expression (log TPM)") +
  theme_minimal_grid() +
  labs(title = glue('*R*<sup>2</sup> = {round(max_corr$r2, 3)}')) +
  theme(plot.title = element_markdown())

fig_corr_grow <- meta_df %>% 
  select(Simple_sample_id, `Growth Rate (1/hr)`) %>% 
  inner_join(., exp_cub_corr, by = c("Simple_sample_id" = "condition")) %>% 
  ggplot() +
    aes(x = `Growth Rate (1/hr)`, y = r2) +
    geom_point(size = 2, alpha = 0.75) +
    scale_y_continuous(name = '*R*<sup>2</sup> (CAI vs expression)') +
    theme_minimal_grid() +
    theme(axis.title.y = element_markdown())

# get spearmans corr
growth_cai_df <- meta_df %>% 
  select(Simple_sample_id, `Growth Rate (1/hr)`) %>% 
  inner_join(., exp_cub_corr, by = c("Simple_sample_id" = "condition"))

cor.test(
  x = growth_cai_df$`Growth Rate (1/hr)`, 
  y = growth_cai_df$r2, 
  method = 'spearman'
)

fig_gr_hist <- meta_df %>% 
  ggplot(aes(x = `Growth Rate (1/hr)`)) +
  geom_histogram(binwidth = 0.05) +
  scale_x_continuous(
    limits = c(0, 1.5),
    expand = c(0,0)
  ) +
  scale_y_continuous(
    name = "Count", 
    expand = c(0,0)
  ) +
  theme_minimal_hgrid()

fig_3 <- plot_grid(
  fig_corr_min, fig_corr_max, fig_gr_hist, fig_corr_grow,
  nrow = 2,
  labels = "AUTO"
) +
  theme(plot.background = element_rect(fill="white", color = NA))

fig_3

rm(fig_corr_max, fig_corr_min, fig_corr_grow, fig_gr_hist,
   min_corr, max_corr, growth_cai_df, corr_df)

```


```{r}

# model with average expression
lm_exp <- lm(CAI ~ mean, data = full_df)

# model with growth correlation
lm_gci <- lm(CAI ~ lin_r, data = full_df)

# model with average expression and gci
lm_exp_gci <- lm(CAI ~ mean + lin_r, data = full_df)

# model with average expression and gci with interaction
lm_inter <- lm(CAI ~ mean + lin_r + mean*lin_r, data = full_df)

fig_5a <- tibble(
  model = c(
    "expression",
    "gci",
    "both_add",
    "both_inter"
  ),
  adj_r2 = c(
    glance(lm_exp)$adj.r.squared,
    glance(lm_gci)$adj.r.squared,
    glance(lm_exp_gci)$adj.r.squared,
    glance(lm_inter)$adj.r.squared
  )
) %>% 
  ggplot(
    aes(
      adj_r2,
      fct_relevel(model, "both_inter", "both_add", "gci", "expression")
    )
  ) +
  geom_col() +
  scale_x_continuous(
    name = "Adjusted *R*<sup>2</sup>",
    limits = c(0, .4)
  ) +
  scale_y_discrete(
    labels = c("Both (interaction)", 
               "Both", "GCI", "Expression")
  ) +
  theme_minimal(base_size = 14) +
  theme(
    axis.title.y = element_blank(),
    axis.title.x = element_markdown()
  )

# create model fits for range of expression and GCI values for use in plotting
new_data <- expand_grid(
  mean = seq(
    from = max(full_df$mean), 
    to = min(full_df$mean), 
    length.out = 100
  ),
  lin_r = seq(
    from = 1, 
    to = -1, 
    length.out = 7)
 )

gci_models <- new_data %>% 
  mutate(
    pred_cai_lm_exp = predict(lm_exp, new_data),
    pred_cai_lm_both = predict(lm_exp_gci, new_data),
    pred_cai_lm_inter = predict(lm_inter, new_data),
    lin_r = factor(round(lin_r, 3))
  )

x <- 2
y <- .8
lab <- "Expression (mean, log TPM)"

# function to create plot 5b expression model
plot_exp_lm <- function(r2_x, r2_y, x_lab) {

  # plot expression model
  tibble(
    CAI = full_df$CAI,
    express = full_df$mean
  ) %>% 
    ggplot() +
    geom_point(aes(express, CAI), alpha = 0.1) +
    geom_line(
      data = gci_models, 
      aes(mean, pred_cai_lm_exp), 
      color = "#CC4678FF", # use color of GCI=0 in other plots
      size = 1
    ) +
    scale_x_continuous(name = x_lab) +
    scale_color_viridis_d(option = "C") +
    theme_half_open() +
    background_grid()
  
} 

# function to create plot 5c model with expression & gci (no interaction)
plot_both_lm <- function(r2_x, r2_y, x_lab) {
  
  # plot additive model
  tibble(
    CAI = full_df$CAI,
    express = full_df$mean
  ) %>% 
    ggplot() +
    geom_point(aes(express, CAI), alpha = 0.1) +
    geom_line(
      data = gci_models, 
      aes(mean, pred_cai_lm_both, color = fct_rev(lin_r)), 
      size = 1
    ) +
    scale_x_continuous(name = x_lab) +
    scale_color_viridis_d(
      option = "C",
      name = "GCI"
    ) +
    theme_half_open() +
    background_grid() +
    theme(legend.position = "none")

}


# function to create plot 5d model with expression & gci with interaction
plot_inter_lm <- function(r2_x, r2_y, x_lab) {
  
  # plot additive model
  tibble(
    CAI = full_df$CAI,
    express = full_df$mean
  ) %>% 
    ggplot() +
    geom_point(aes(express, CAI), alpha = 0.1) +
    geom_line(
      data = gci_models, 
      aes(mean, pred_cai_lm_inter, color = fct_rev(lin_r)), 
      size = 1
    ) +
    scale_x_continuous(name = x_lab) +
    scale_color_viridis_d(
      option = "C",
      name = "GCI"
    ) +
    theme_half_open() +
    background_grid() +
    theme(legend.position = "none")

}

# create plots
fig_5b <- plot_exp_lm(x, y, lab)

fig_5c <- plot_both_lm(x, y, lab)

fig_5d <- plot_inter_lm(x, y, lab)

# extract shared legend
legend <- get_legend(
  fig_5c + 
    guides(color = guide_legend(nrow = 1)) +
    theme(legend.position = "bottom")
)

plots <- plot_grid(
  fig_5a, fig_5b, fig_5c, fig_5d,
  nrow = 2,
  labels = "AUTO"
) 

fig_5 <- plot_grid(
  plots, legend,
  nrow = 2,
  rel_heights = c(1, .05)
) +
  theme(plot.background = element_rect(fill="white", color = NA))

fig_5


```



```{r}

save_plot(
  here(
    "Manuscript", 
    "fig_CAI_exp_gr_no_evol.png"
  ),
  fig_3, 
  base_height = 6.71,
  base_asp = 1.618, 
  base_width = NULL
)

save_plot(
  here(
    "Manuscript", 
    "fig_CAI_gci_model_no_evol.png" 
  ),
  fig_5, 
  ncol = 1, 
  nrow = 1, 
  base_height = 6.71, #base_height = 5.71,
  base_asp = 1.618, 
  base_width = NULL
)

rm(fig_5a, fig_5b, fig_5c, fig_5d, fig_5, fig_3, gci_models, legend, lm_exp, lm_exp_gci, lm_gci, lm_inter, new_data, plots, y, x, lab, plot_both_lm, plot_exp_lm, plot_inter_lm)

```


# 6 - go and core analysis
```{r}

genes_positive <- exp_gr_corr %>%
  filter(pearsons > 0) %>%
  select(genes, pearsons) %>%
  left_join(., full_df, by = c("genes" = "locus_tag"))
colnames(genes_positive) <- c("locus_tag", "GCI", "gene_name")

genes_negative <- exp_gr_corr %>%
  filter(pearsons < 0) %>%
  select(genes, pearsons) %>%
  left_join(., full_df, by = c("genes" = "locus_tag"))
colnames(genes_negative) <- c("locus_tag", "GCI", "gene_name")

rm(exp_cub_corr, exp_df, exp_gr_corr, exp_gr_df, full_df, meta_df)

library(AnnotationDbi)
library(org.EcK12.eg.db)
library(clusterProfiler)

data<- genes_positive %>% 
  select(locus_tag, GCI, gene_name)
neg_data <- genes_negative%>% 
  select(locus_tag, GCI, gene_name)

data$ENTREZID <- mapIds(org.EcK12.eg.db,
                              keys = data$gene_name,
                              column = "ENTREZID",
                              keytype = "SYMBOL",
                              multiVals = "first")

neg_data$ENTREZID <- mapIds(org.EcK12.eg.db,
                              keys = neg_data$gene_name,
                              column = "ENTREZID",
                              keytype = "SYMBOL",
                              multiVals = "first")

# removing genes without entrezID
mtx <- subset(data, is.na(ENTREZID) == FALSE)
n_mtx <- subset(neg_data, is.na(ENTREZID) == FALSE)

# creating a new matrix
gene_matrix <- mtx$GCI
n_gene_matrix <- n_mtx$GCI

# add entrezID as rownames 
names(gene_matrix) <- mtx$ENTREZID
names(n_gene_matrix) <- n_mtx$ENTREZID

# subseting GCI > 0
# in this case nothing is done
geneList <- gene_matrix[abs(gene_matrix) > 0] 
n_geneList <- n_gene_matrix[abs(n_gene_matrix) > 0] 

go_enrich_ALL <- enrichGO(
  gene=names(geneList),
  OrgDb= 'org.EcK12.eg.db',
  keyType = "ENTREZID",
  ont = "ALL",
  pvalueCutoff = 0.04,
  pAdjustMethod = "BH",
  qvalueCutoff = 0.05,
  readable = TRUE,
  pool = FALSE
)

summary_go_ALL <- as.data.frame(go_enrich_ALL)

go_enrich_ALL_neg <- enrichGO(
  gene=names(n_geneList),
  OrgDb= 'org.EcK12.eg.db',
  keyType = "ENTREZID",
  ont = "ALL",
  pvalueCutoff = 0.04,
  pAdjustMethod = "BH",
  qvalueCutoff = 0.05,
  readable = TRUE,
  pool = FALSE
)

summary_go_ALL_neg <- as.data.frame(go_enrich_ALL_neg)

df_go_pos_gci <- summary_go_ALL %>% 
  select(c(Description, GeneRatio, BgRatio, p.adjust)) %>% # keep only necessary columns
  extract( # split ratios into numerator and denominator to build contingency table
    GeneRatio, 
    into = c("gene_num", "gene_denom"), 
    regex = "([[:alnum:]]+)/([[:alnum:]]+)"
  ) %>% 
  extract(
    BgRatio,
    into = c("bg_num", "bg_denom"),
    regex = "([[:alnum:]]+)/([[:alnum:]]+)"
  ) %>% 
  dplyr::mutate(
    cat_a = as.numeric(gene_num),
    cat_b = as.numeric(gene_denom) - cat_a,
    cat_c = as.numeric(bg_num) - cat_a,
    cat_d = as.numeric(bg_denom) - cat_a - cat_b - cat_c,
    odds_ratio = (cat_a/cat_b)/(cat_c/cat_d)
  ) 
  
df_go_neg_gci <- summary_go_ALL_neg %>% 
  select(c(Description, GeneRatio, BgRatio, p.adjust)) %>% # keep only necessary columns
  extract( # split ratios into numerator and denominator to build contingency table
    GeneRatio, 
    into = c("gene_num", "gene_denom"), 
    regex = "([[:alnum:]]+)/([[:alnum:]]+)"
  ) %>% 
  extract(
    BgRatio,
    into = c("bg_num", "bg_denom"),
    regex = "([[:alnum:]]+)/([[:alnum:]]+)"
  ) %>% 
  dplyr::mutate(
    cat_a = as.numeric(gene_num),
    cat_b = as.numeric(gene_denom) - cat_a,
    cat_c = as.numeric(bg_num) - cat_a,
    cat_d = as.numeric(bg_denom) - cat_a - cat_b - cat_c,
    odds_ratio = (cat_a/cat_b)/(cat_c/cat_d)
  )

fig_go_pos <- df_go_pos_gci %>% 
  arrange(desc(odds_ratio)) %>% 
  slice(1:15) %>% 
  mutate(labels = str_wrap(Description, 30)) %>% 
  ggplot() +
  aes(x = odds_ratio, y = fct_reorder(labels, odds_ratio)) +
  geom_point(size = 3) +
  geom_vline(xintercept = 1) +
  #scale_x_continuous(name = "Odds Ratio") +
  scale_x_log10(name = "Odds Ratio", limits = c(1, 40)) +
  scale_y_discrete(name = NULL) +
  ggtitle("GO enrichment (GCI > 0)") +
  theme_bw()

fig_go_neg <- df_go_neg_gci %>% 
  arrange(desc(odds_ratio)) %>% 
  slice(1:15) %>% 
  mutate(
    labels = str_wrap(Description, 30)
  ) %>% 
  ggplot() +
  geom_point(
    aes(
      x = odds_ratio, 
      y = fct_reorder(labels, odds_ratio)
    ),
    size = 3
  ) +
  geom_vline(xintercept = 1) +
  #scale_x_continuous(name = "Odds Ratio") +
  scale_x_log10(name = "Odds Ratio", limits = c(1, 40)) +
  scale_y_discrete(name = NULL) +
  ggtitle("GO enrichment (GCI < 0)") +
  theme_bw() 

rm(data, neg_data, mtx, n_mtx, go_enrich_ALL, go_enrich_ALL_neg)

# essential gene data set with unknown Adam origins
core_df1 <- read_tsv(
  file = here('Data', 'ecoli_info', 'essential_ecoli_genes.txt')
) %>% 
  select(`Gene Name`, `Note`)

# add gene classifications into data set
full_df <- full_df %>% 
  mutate(
    core1 = case_when(
      gene %in% core_df1$`Gene Name` ~ "essential",
      TRUE ~ "non-essential"
    )
  ) 

exp_gr_df <- exp_df %>% 
    gather(-`log-TPM`,
           key = "condition",
           value = "expression") %>% 
    spread(`log-TPM`, expression)

exp_gr_df <- meta_df %>% 
    select(`Simple_sample_id`, `Growth Rate (1/hr)`) %>% 
    inner_join(., exp_gr_df, by = c("Simple_sample_id" = "condition"))

exp_gr_corr <- cor(
  exp_gr_df[2], 
  exp_gr_df[3:ncol(exp_gr_df)], 
  method = "pearson"
) %>% 
  as_tibble() %>% 
  gather(
    .,
    key = "genes",
    value = "pearsons"
  ) %>%
  mutate(r2 = pearsons^2)

exp_gr_corr$genes <- gsub("_[0-9]", "", exp_gr_corr$genes)

# join all data for gene classification and gci values
gci_genomes_df <- exp_gr_corr %>% 
  left_join(., full_df, by = c("genes" = "locus_tag"))

# data set 1
core_gci1 <- gci_genomes_df %>% 
  filter(core1 == "essential") %>% 
  select(pearsons)

access_gci1 <- gci_genomes_df %>% 
  filter(core1 == "non-essential") %>% 
  select(pearsons)

t_test_results <- t.test(core_gci1, access_gci1)
t_test_results$p.value  # 9.292282e-58

# use essential data set from Adam in figure
# create label data set
ttest_text <- data.frame(
  label = glue::glue(
    'p-value = {scales::scientific(
    t_test_results$p.value, digits = 3
    )}'),
  x = 0.3, 
  y = 1.4
)

fig_core_gci <- gci_genomes_df %>% 
  drop_na(core1) %>% 
  ggplot(aes(x = pearsons, fill = core1)) +
  geom_density(alpha = 0.5) +
  geom_richtext(
    data = ttest_text,
    aes(x = x, y = y, label = label),
    label.color = NA,
    inherit.aes = FALSE
  ) +
  scale_x_continuous(name = "GCI") +
  scale_y_continuous(name = "Density") +
  scale_fill_viridis_d(name = "Gene type") +
  theme_minimal_grid()

fig_go_plots <- plot_grid(
  fig_go_neg, fig_go_pos,
  nrow = 1
)

fig_6 <- plot_grid(
  fig_core_gci, fig_go_plots,
  nrow = 2,
  rel_heights = c(.5, 1),
  labels = "AUTO"
) + 
  theme(plot.background = element_rect(fill="white", color = NA))

fig_6

save_plot(
  here('Manuscript', 'fig_core_go_gci.png'),
  fig_6, 
  ncol = 1, 
  nrow = 1, 
  base_height = 6.71, #base_height = 5.71,
  base_asp = 1.618, 
  base_width = NULL
)


```





