{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy import stats\n",
    "\n",
    "import statsmodels.formula.api as smf\n",
    "\n",
    "from matplotlib import pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This data comes from this paper: https://www.nature.com/articles/s41467-019-13483-w\n",
    "\n",
    "It's a big excel file with expression and growth rate data in different sheets and this code will use a few of these sheets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.ExcelFile(r'../Data/raw_data/41467_2019_13483_MOESM4_ESM.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Basic metadata clean up and data subsetting\n",
    "\n",
    "After some manual inspection, looking at distributions, etc. we decided to exclude samples where:\n",
    "\n",
    "1. The growth rate data was unknown\n",
    "\n",
    "\n",
    "2. The growth rate data was reported as zero. This one might seem strange but it's is a little unclear in general if that's possible/true to have zero growth rate. It is possible that these were stationary phase cultures but equally likely from my stand-point that these are errors in the table.\n",
    "\n",
    "\n",
    "3. Really poor alignment (perhaps indicating some overall contamination)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(278, 26)\n",
      "(195, 26)\n",
      "(179, 26)\n",
      "(173, 26)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Study</th>\n",
       "      <th>Project ID</th>\n",
       "      <th>Condition ID</th>\n",
       "      <th>Replicate #</th>\n",
       "      <th>Strain Description</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Base Media</th>\n",
       "      <th>Carbon Source (g/L)</th>\n",
       "      <th>Nitrogen Source (g/L)</th>\n",
       "      <th>...</th>\n",
       "      <th>Culture Type</th>\n",
       "      <th>Growth Rate (1/hr)</th>\n",
       "      <th>Evolved Sample</th>\n",
       "      <th>Isolate Type</th>\n",
       "      <th>Sequencing Machine</th>\n",
       "      <th>Additional Details</th>\n",
       "      <th>Biological Replicates</th>\n",
       "      <th>Alignment</th>\n",
       "      <th>DOI</th>\n",
       "      <th>GEO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>fur__wt_fe__1</td>\n",
       "      <td>Fur</td>\n",
       "      <td>fur</td>\n",
       "      <td>wt_fe</td>\n",
       "      <td>1</td>\n",
       "      <td>Escherichia coli K-12 MG1655</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>...</td>\n",
       "      <td>Batch</td>\n",
       "      <td>1.060606</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>93.35</td>\n",
       "      <td>doi.org/10.1038/ncomms5910</td>\n",
       "      <td>GSE54900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>fur__wt_fe__2</td>\n",
       "      <td>Fur</td>\n",
       "      <td>fur</td>\n",
       "      <td>wt_fe</td>\n",
       "      <td>2</td>\n",
       "      <td>Escherichia coli K-12 MG1655</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>...</td>\n",
       "      <td>Batch</td>\n",
       "      <td>1.060606</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>92.38</td>\n",
       "      <td>doi.org/10.1038/ncomms5910</td>\n",
       "      <td>GSE54900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>fur__delfur_fe2__1</td>\n",
       "      <td>Fur</td>\n",
       "      <td>fur</td>\n",
       "      <td>delfur_fe2</td>\n",
       "      <td>1</td>\n",
       "      <td>Escherichia coli K-12 MG1655 del_fur</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>...</td>\n",
       "      <td>Batch</td>\n",
       "      <td>0.619469</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>92.80</td>\n",
       "      <td>doi.org/10.1038/ncomms5910</td>\n",
       "      <td>GSE54900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>fur__delfur_fe2__2</td>\n",
       "      <td>Fur</td>\n",
       "      <td>fur</td>\n",
       "      <td>delfur_fe2</td>\n",
       "      <td>2</td>\n",
       "      <td>Escherichia coli K-12 MG1655 del_fur</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>...</td>\n",
       "      <td>Batch</td>\n",
       "      <td>0.619469</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>93.24</td>\n",
       "      <td>doi.org/10.1038/ncomms5910</td>\n",
       "      <td>GSE54900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>omics__bw_ac__1</td>\n",
       "      <td>Omics</td>\n",
       "      <td>omics</td>\n",
       "      <td>bw_ac</td>\n",
       "      <td>1</td>\n",
       "      <td>Escherichia coli BW25113</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>M9</td>\n",
       "      <td>acetate(3.5)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>...</td>\n",
       "      <td>Batch</td>\n",
       "      <td>0.203000</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>97.80</td>\n",
       "      <td>doi.org/10.1038/ncomms13091</td>\n",
       "      <td>GSE59759</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Sample ID  Study Project ID Condition ID  Replicate #  \\\n",
       "4        fur__wt_fe__1    Fur        fur        wt_fe            1   \n",
       "5        fur__wt_fe__2    Fur        fur        wt_fe            2   \n",
       "8   fur__delfur_fe2__1    Fur        fur   delfur_fe2            1   \n",
       "9   fur__delfur_fe2__2    Fur        fur   delfur_fe2            2   \n",
       "55     omics__bw_ac__1  Omics      omics        bw_ac            1   \n",
       "\n",
       "                      Strain Description   Strain Base Media  \\\n",
       "4           Escherichia coli K-12 MG1655   MG1655         M9   \n",
       "5           Escherichia coli K-12 MG1655   MG1655         M9   \n",
       "8   Escherichia coli K-12 MG1655 del_fur   MG1655         M9   \n",
       "9   Escherichia coli K-12 MG1655 del_fur   MG1655         M9   \n",
       "55              Escherichia coli BW25113  BW25113         M9   \n",
       "\n",
       "   Carbon Source (g/L) Nitrogen Source (g/L)  ... Culture Type  \\\n",
       "4           glucose(2)              NH4Cl(1)  ...        Batch   \n",
       "5           glucose(2)              NH4Cl(1)  ...        Batch   \n",
       "8           glucose(2)              NH4Cl(1)  ...        Batch   \n",
       "9           glucose(2)              NH4Cl(1)  ...        Batch   \n",
       "55        acetate(3.5)              NH4Cl(1)  ...        Batch   \n",
       "\n",
       "   Growth Rate (1/hr) Evolved Sample  Isolate Type  Sequencing Machine  \\\n",
       "4            1.060606             No           NaN               MiSeq   \n",
       "5            1.060606             No           NaN               MiSeq   \n",
       "8            0.619469             No           NaN               MiSeq   \n",
       "9            0.619469             No           NaN               MiSeq   \n",
       "55           0.203000             No           NaN               MiSeq   \n",
       "\n",
       "   Additional Details Biological Replicates  Alignment  \\\n",
       "4                 NaN                     2      93.35   \n",
       "5                 NaN                     2      92.38   \n",
       "8                 NaN                     2      92.80   \n",
       "9                 NaN                     2      93.24   \n",
       "55                NaN                     2      97.80   \n",
       "\n",
       "                            DOI       GEO  \n",
       "4    doi.org/10.1038/ncomms5910  GSE54900  \n",
       "5    doi.org/10.1038/ncomms5910  GSE54900  \n",
       "8    doi.org/10.1038/ncomms5910  GSE54900  \n",
       "9    doi.org/10.1038/ncomms5910  GSE54900  \n",
       "55  doi.org/10.1038/ncomms13091  GSE59759  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_df = df.parse('Metadata') ###This grabs the sheet that contains information about the samples\n",
    "print(meta_df.shape)\n",
    "meta_df = meta_df[meta_df['Growth Rate (1/hr)'].isnull() == False]\n",
    "print(meta_df.shape)\n",
    "meta_df = meta_df[meta_df['Growth Rate (1/hr)'] > 0.0]\n",
    "print(meta_df.shape)\n",
    "meta_df = meta_df[meta_df['Alignment'] > 80]\n",
    "print(meta_df.shape)\n",
    "meta_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in the expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3923, 278)\n",
      "(3923, 173)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fur__wt_fe__1</th>\n",
       "      <th>fur__wt_fe__2</th>\n",
       "      <th>fur__delfur_fe2__1</th>\n",
       "      <th>fur__delfur_fe2__2</th>\n",
       "      <th>omics__bw_ac__1</th>\n",
       "      <th>omics__bw_ac__2</th>\n",
       "      <th>omics__bw_fum__1</th>\n",
       "      <th>omics__bw_fum__2</th>\n",
       "      <th>omics__bw_glc__1</th>\n",
       "      <th>omics__bw_glc__2</th>\n",
       "      <th>...</th>\n",
       "      <th>efeU__menFentC_ale29__1</th>\n",
       "      <th>efeU__menFentC_ale29__2</th>\n",
       "      <th>efeU__menFentC_ale30__1</th>\n",
       "      <th>efeU__menFentC_ale30__2</th>\n",
       "      <th>efeU__menFentCubiC_ale36__1</th>\n",
       "      <th>efeU__menFentCubiC_ale36__2</th>\n",
       "      <th>efeU__menFentCubiC_ale37__1</th>\n",
       "      <th>efeU__menFentCubiC_ale37__2</th>\n",
       "      <th>efeU__menFentCubiC_ale38__1</th>\n",
       "      <th>efeU__menFentCubiC_ale38__2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log-TPM</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>11.064810</td>\n",
       "      <td>10.779071</td>\n",
       "      <td>11.229767</td>\n",
       "      <td>11.214065</td>\n",
       "      <td>9.257348</td>\n",
       "      <td>9.182322</td>\n",
       "      <td>9.709213</td>\n",
       "      <td>9.672126</td>\n",
       "      <td>10.208587</td>\n",
       "      <td>10.218351</td>\n",
       "      <td>...</td>\n",
       "      <td>10.271327</td>\n",
       "      <td>10.276565</td>\n",
       "      <td>11.148538</td>\n",
       "      <td>11.170578</td>\n",
       "      <td>11.676604</td>\n",
       "      <td>11.726097</td>\n",
       "      <td>11.881529</td>\n",
       "      <td>11.923237</td>\n",
       "      <td>11.496410</td>\n",
       "      <td>11.552762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>10.776984</td>\n",
       "      <td>10.597810</td>\n",
       "      <td>10.897938</td>\n",
       "      <td>10.861157</td>\n",
       "      <td>8.983408</td>\n",
       "      <td>8.943151</td>\n",
       "      <td>9.436004</td>\n",
       "      <td>9.394573</td>\n",
       "      <td>9.609637</td>\n",
       "      <td>9.677931</td>\n",
       "      <td>...</td>\n",
       "      <td>10.160291</td>\n",
       "      <td>10.116861</td>\n",
       "      <td>10.314322</td>\n",
       "      <td>10.392251</td>\n",
       "      <td>10.916426</td>\n",
       "      <td>10.909277</td>\n",
       "      <td>11.023924</td>\n",
       "      <td>11.038426</td>\n",
       "      <td>10.624301</td>\n",
       "      <td>10.764195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>10.394971</td>\n",
       "      <td>10.113950</td>\n",
       "      <td>10.185151</td>\n",
       "      <td>10.164655</td>\n",
       "      <td>8.761690</td>\n",
       "      <td>8.779920</td>\n",
       "      <td>9.532673</td>\n",
       "      <td>9.534370</td>\n",
       "      <td>9.883558</td>\n",
       "      <td>9.870356</td>\n",
       "      <td>...</td>\n",
       "      <td>10.475069</td>\n",
       "      <td>10.434352</td>\n",
       "      <td>10.679541</td>\n",
       "      <td>10.723953</td>\n",
       "      <td>11.143310</td>\n",
       "      <td>11.112721</td>\n",
       "      <td>11.184795</td>\n",
       "      <td>11.241845</td>\n",
       "      <td>10.953206</td>\n",
       "      <td>11.001006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>6.716069</td>\n",
       "      <td>6.410864</td>\n",
       "      <td>6.527653</td>\n",
       "      <td>6.136168</td>\n",
       "      <td>4.474204</td>\n",
       "      <td>4.720490</td>\n",
       "      <td>5.782102</td>\n",
       "      <td>5.326669</td>\n",
       "      <td>5.846675</td>\n",
       "      <td>5.972022</td>\n",
       "      <td>...</td>\n",
       "      <td>5.979079</td>\n",
       "      <td>5.705586</td>\n",
       "      <td>6.306120</td>\n",
       "      <td>6.291340</td>\n",
       "      <td>5.058537</td>\n",
       "      <td>4.835550</td>\n",
       "      <td>5.448097</td>\n",
       "      <td>5.757951</td>\n",
       "      <td>5.873964</td>\n",
       "      <td>5.808618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "      <td>6.761813</td>\n",
       "      <td>6.816532</td>\n",
       "      <td>6.862147</td>\n",
       "      <td>6.817480</td>\n",
       "      <td>6.536457</td>\n",
       "      <td>6.439917</td>\n",
       "      <td>6.408731</td>\n",
       "      <td>6.276017</td>\n",
       "      <td>6.910200</td>\n",
       "      <td>6.843384</td>\n",
       "      <td>...</td>\n",
       "      <td>8.371287</td>\n",
       "      <td>8.322390</td>\n",
       "      <td>8.137515</td>\n",
       "      <td>8.071837</td>\n",
       "      <td>7.354131</td>\n",
       "      <td>7.365536</td>\n",
       "      <td>7.328101</td>\n",
       "      <td>7.314761</td>\n",
       "      <td>8.057750</td>\n",
       "      <td>8.105213</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 173 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         fur__wt_fe__1  fur__wt_fe__2  fur__delfur_fe2__1  fur__delfur_fe2__2  \\\n",
       "log-TPM                                                                         \n",
       "b0002        11.064810      10.779071           11.229767           11.214065   \n",
       "b0003        10.776984      10.597810           10.897938           10.861157   \n",
       "b0004        10.394971      10.113950           10.185151           10.164655   \n",
       "b0005         6.716069       6.410864            6.527653            6.136168   \n",
       "b0006         6.761813       6.816532            6.862147            6.817480   \n",
       "\n",
       "         omics__bw_ac__1  omics__bw_ac__2  omics__bw_fum__1  omics__bw_fum__2  \\\n",
       "log-TPM                                                                         \n",
       "b0002           9.257348         9.182322          9.709213          9.672126   \n",
       "b0003           8.983408         8.943151          9.436004          9.394573   \n",
       "b0004           8.761690         8.779920          9.532673          9.534370   \n",
       "b0005           4.474204         4.720490          5.782102          5.326669   \n",
       "b0006           6.536457         6.439917          6.408731          6.276017   \n",
       "\n",
       "         omics__bw_glc__1  omics__bw_glc__2  ...  efeU__menFentC_ale29__1  \\\n",
       "log-TPM                                      ...                            \n",
       "b0002           10.208587         10.218351  ...                10.271327   \n",
       "b0003            9.609637          9.677931  ...                10.160291   \n",
       "b0004            9.883558          9.870356  ...                10.475069   \n",
       "b0005            5.846675          5.972022  ...                 5.979079   \n",
       "b0006            6.910200          6.843384  ...                 8.371287   \n",
       "\n",
       "         efeU__menFentC_ale29__2  efeU__menFentC_ale30__1  \\\n",
       "log-TPM                                                     \n",
       "b0002                  10.276565                11.148538   \n",
       "b0003                  10.116861                10.314322   \n",
       "b0004                  10.434352                10.679541   \n",
       "b0005                   5.705586                 6.306120   \n",
       "b0006                   8.322390                 8.137515   \n",
       "\n",
       "         efeU__menFentC_ale30__2  efeU__menFentCubiC_ale36__1  \\\n",
       "log-TPM                                                         \n",
       "b0002                  11.170578                    11.676604   \n",
       "b0003                  10.392251                    10.916426   \n",
       "b0004                  10.723953                    11.143310   \n",
       "b0005                   6.291340                     5.058537   \n",
       "b0006                   8.071837                     7.354131   \n",
       "\n",
       "         efeU__menFentCubiC_ale36__2  efeU__menFentCubiC_ale37__1  \\\n",
       "log-TPM                                                             \n",
       "b0002                      11.726097                    11.881529   \n",
       "b0003                      10.909277                    11.023924   \n",
       "b0004                      11.112721                    11.184795   \n",
       "b0005                       4.835550                     5.448097   \n",
       "b0006                       7.365536                     7.328101   \n",
       "\n",
       "         efeU__menFentCubiC_ale37__2  efeU__menFentCubiC_ale38__1  \\\n",
       "log-TPM                                                             \n",
       "b0002                      11.923237                    11.496410   \n",
       "b0003                      11.038426                    10.624301   \n",
       "b0004                      11.241845                    10.953206   \n",
       "b0005                       5.757951                     5.873964   \n",
       "b0006                       7.314761                     8.057750   \n",
       "\n",
       "         efeU__menFentCubiC_ale38__2  \n",
       "log-TPM                               \n",
       "b0002                      11.552762  \n",
       "b0003                      10.764195  \n",
       "b0004                      11.001006  \n",
       "b0005                       5.808618  \n",
       "b0006                       8.105213  \n",
       "\n",
       "[5 rows x 173 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exp_df = df.parse('Expression Data', index_col='log-TPM')\n",
    "print(exp_df.shape)\n",
    "exp_df = exp_df[meta_df['Sample ID']] ###Only grab the columns corresponding to the samples identified above\n",
    "print(exp_df.shape)\n",
    "assert list(meta_df['Sample ID']) == list(exp_df.columns) ###Check our work\n",
    "exp_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Manual inspection found some weird and highly implausible/impossible duplicate column/s, so we'll make a note of that here and deal with it later**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pal__lyx_ale2_f6__1</th>\n",
       "      <th>pal__lyx_ale2__1</th>\n",
       "      <th>pal__lyx_ale2__2</th>\n",
       "      <th>pal__lyx_ale4__1</th>\n",
       "      <th>pal__lyx_ale4__2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log-TPM</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>9.627287</td>\n",
       "      <td>10.130315</td>\n",
       "      <td>10.155462</td>\n",
       "      <td>10.130315</td>\n",
       "      <td>10.155462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>9.250534</td>\n",
       "      <td>9.708944</td>\n",
       "      <td>9.831138</td>\n",
       "      <td>9.708944</td>\n",
       "      <td>9.831138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>9.203814</td>\n",
       "      <td>10.049444</td>\n",
       "      <td>10.190627</td>\n",
       "      <td>10.049444</td>\n",
       "      <td>10.190627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>4.807384</td>\n",
       "      <td>5.772047</td>\n",
       "      <td>5.933463</td>\n",
       "      <td>5.772047</td>\n",
       "      <td>5.933463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "      <td>6.398236</td>\n",
       "      <td>6.435048</td>\n",
       "      <td>6.220552</td>\n",
       "      <td>6.435048</td>\n",
       "      <td>6.220552</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         pal__lyx_ale2_f6__1  pal__lyx_ale2__1  pal__lyx_ale2__2  \\\n",
       "log-TPM                                                            \n",
       "b0002               9.627287         10.130315         10.155462   \n",
       "b0003               9.250534          9.708944          9.831138   \n",
       "b0004               9.203814         10.049444         10.190627   \n",
       "b0005               4.807384          5.772047          5.933463   \n",
       "b0006               6.398236          6.435048          6.220552   \n",
       "\n",
       "         pal__lyx_ale4__1  pal__lyx_ale4__2  \n",
       "log-TPM                                      \n",
       "b0002           10.130315         10.155462  \n",
       "b0003            9.708944          9.831138  \n",
       "b0004           10.049444         10.190627  \n",
       "b0005            5.772047          5.933463  \n",
       "b0006            6.435048          6.220552  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exp_df[[col for col in exp_df.columns if 'pal__lyx_ale' in col]].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Averaging gene expression values across replicates to clean up/simplify the data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## First getting gene expression averages between replicates and creating a new `dataframe` to hold this information\n",
    "\n",
    "Thus just exploits the fact that replicates are denoted by \"__x\" in the sample names so we identify these and average them when possible. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3923, 105)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_23143/3440901807.py:5: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  new_exp_df[i] = exp_df[[col for col in exp_df.columns if col[:-3]==i]].mean(axis=1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pal__arab_ale14</th>\n",
       "      <th>ica__thm_gal</th>\n",
       "      <th>ssw__xyl_ale3</th>\n",
       "      <th>fur__delfur_fe2</th>\n",
       "      <th>ytf__wt_ph5</th>\n",
       "      <th>42c__42c_ale8</th>\n",
       "      <th>ytf__delyheO</th>\n",
       "      <th>cra_crp__delcra_glc</th>\n",
       "      <th>cra_crp__delcra_ac</th>\n",
       "      <th>pgi__pgi_ale1</th>\n",
       "      <th>...</th>\n",
       "      <th>rpoB__rpoBE546V_lb</th>\n",
       "      <th>ssw__wt_glyc</th>\n",
       "      <th>ica__gth</th>\n",
       "      <th>glu__glu_ale4</th>\n",
       "      <th>efeU__menFentCubiC_ale36</th>\n",
       "      <th>fps__fps_ptsI_ale1</th>\n",
       "      <th>42c__42c_ale2</th>\n",
       "      <th>omics__bw_ac</th>\n",
       "      <th>ytf__delyeiE</th>\n",
       "      <th>pgi__pgi_ale2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log-TPM</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>10.504998</td>\n",
       "      <td>11.090470</td>\n",
       "      <td>9.865774</td>\n",
       "      <td>11.221916</td>\n",
       "      <td>11.069910</td>\n",
       "      <td>10.997682</td>\n",
       "      <td>9.401156</td>\n",
       "      <td>11.230252</td>\n",
       "      <td>8.487038</td>\n",
       "      <td>9.816061</td>\n",
       "      <td>...</td>\n",
       "      <td>8.163975</td>\n",
       "      <td>7.678018</td>\n",
       "      <td>8.319211</td>\n",
       "      <td>11.038749</td>\n",
       "      <td>11.701351</td>\n",
       "      <td>9.955873</td>\n",
       "      <td>11.216042</td>\n",
       "      <td>9.219835</td>\n",
       "      <td>10.874816</td>\n",
       "      <td>9.783981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>10.025917</td>\n",
       "      <td>10.402663</td>\n",
       "      <td>9.473220</td>\n",
       "      <td>10.879548</td>\n",
       "      <td>10.408418</td>\n",
       "      <td>10.747250</td>\n",
       "      <td>8.933743</td>\n",
       "      <td>11.106092</td>\n",
       "      <td>8.546123</td>\n",
       "      <td>9.163024</td>\n",
       "      <td>...</td>\n",
       "      <td>8.127386</td>\n",
       "      <td>8.384440</td>\n",
       "      <td>7.711763</td>\n",
       "      <td>10.443008</td>\n",
       "      <td>10.912852</td>\n",
       "      <td>9.536160</td>\n",
       "      <td>10.222039</td>\n",
       "      <td>8.963280</td>\n",
       "      <td>9.861151</td>\n",
       "      <td>8.990193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>10.296601</td>\n",
       "      <td>10.633509</td>\n",
       "      <td>9.494152</td>\n",
       "      <td>10.174903</td>\n",
       "      <td>10.535924</td>\n",
       "      <td>10.247703</td>\n",
       "      <td>9.164724</td>\n",
       "      <td>10.486687</td>\n",
       "      <td>7.748146</td>\n",
       "      <td>9.134662</td>\n",
       "      <td>...</td>\n",
       "      <td>8.065428</td>\n",
       "      <td>8.745754</td>\n",
       "      <td>7.557831</td>\n",
       "      <td>10.551490</td>\n",
       "      <td>11.128015</td>\n",
       "      <td>9.797032</td>\n",
       "      <td>10.257204</td>\n",
       "      <td>8.770805</td>\n",
       "      <td>9.682288</td>\n",
       "      <td>9.040825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>5.452944</td>\n",
       "      <td>4.686068</td>\n",
       "      <td>5.862383</td>\n",
       "      <td>6.331911</td>\n",
       "      <td>5.676045</td>\n",
       "      <td>6.417454</td>\n",
       "      <td>4.215120</td>\n",
       "      <td>5.818731</td>\n",
       "      <td>3.399998</td>\n",
       "      <td>5.705243</td>\n",
       "      <td>...</td>\n",
       "      <td>3.848184</td>\n",
       "      <td>5.925175</td>\n",
       "      <td>4.999832</td>\n",
       "      <td>6.360466</td>\n",
       "      <td>4.947044</td>\n",
       "      <td>4.861505</td>\n",
       "      <td>5.687336</td>\n",
       "      <td>4.597347</td>\n",
       "      <td>4.786873</td>\n",
       "      <td>4.690402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "      <td>6.514145</td>\n",
       "      <td>6.663905</td>\n",
       "      <td>6.849183</td>\n",
       "      <td>6.839814</td>\n",
       "      <td>7.108237</td>\n",
       "      <td>7.046514</td>\n",
       "      <td>6.748235</td>\n",
       "      <td>6.937212</td>\n",
       "      <td>6.647096</td>\n",
       "      <td>7.159757</td>\n",
       "      <td>...</td>\n",
       "      <td>6.963816</td>\n",
       "      <td>6.186970</td>\n",
       "      <td>6.235536</td>\n",
       "      <td>6.816227</td>\n",
       "      <td>7.359833</td>\n",
       "      <td>6.910678</td>\n",
       "      <td>6.727618</td>\n",
       "      <td>6.488187</td>\n",
       "      <td>7.009695</td>\n",
       "      <td>7.552121</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 105 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         pal__arab_ale14  ica__thm_gal  ssw__xyl_ale3  fur__delfur_fe2  \\\n",
       "log-TPM                                                                  \n",
       "b0002          10.504998     11.090470       9.865774        11.221916   \n",
       "b0003          10.025917     10.402663       9.473220        10.879548   \n",
       "b0004          10.296601     10.633509       9.494152        10.174903   \n",
       "b0005           5.452944      4.686068       5.862383         6.331911   \n",
       "b0006           6.514145      6.663905       6.849183         6.839814   \n",
       "\n",
       "         ytf__wt_ph5  42c__42c_ale8  ytf__delyheO  cra_crp__delcra_glc  \\\n",
       "log-TPM                                                                  \n",
       "b0002      11.069910      10.997682      9.401156            11.230252   \n",
       "b0003      10.408418      10.747250      8.933743            11.106092   \n",
       "b0004      10.535924      10.247703      9.164724            10.486687   \n",
       "b0005       5.676045       6.417454      4.215120             5.818731   \n",
       "b0006       7.108237       7.046514      6.748235             6.937212   \n",
       "\n",
       "         cra_crp__delcra_ac  pgi__pgi_ale1  ...  rpoB__rpoBE546V_lb  \\\n",
       "log-TPM                                     ...                       \n",
       "b0002              8.487038       9.816061  ...            8.163975   \n",
       "b0003              8.546123       9.163024  ...            8.127386   \n",
       "b0004              7.748146       9.134662  ...            8.065428   \n",
       "b0005              3.399998       5.705243  ...            3.848184   \n",
       "b0006              6.647096       7.159757  ...            6.963816   \n",
       "\n",
       "         ssw__wt_glyc  ica__gth  glu__glu_ale4  efeU__menFentCubiC_ale36  \\\n",
       "log-TPM                                                                    \n",
       "b0002        7.678018  8.319211      11.038749                 11.701351   \n",
       "b0003        8.384440  7.711763      10.443008                 10.912852   \n",
       "b0004        8.745754  7.557831      10.551490                 11.128015   \n",
       "b0005        5.925175  4.999832       6.360466                  4.947044   \n",
       "b0006        6.186970  6.235536       6.816227                  7.359833   \n",
       "\n",
       "         fps__fps_ptsI_ale1  42c__42c_ale2  omics__bw_ac  ytf__delyeiE  \\\n",
       "log-TPM                                                                  \n",
       "b0002              9.955873      11.216042      9.219835     10.874816   \n",
       "b0003              9.536160      10.222039      8.963280      9.861151   \n",
       "b0004              9.797032      10.257204      8.770805      9.682288   \n",
       "b0005              4.861505       5.687336      4.597347      4.786873   \n",
       "b0006              6.910678       6.727618      6.488187      7.009695   \n",
       "\n",
       "         pgi__pgi_ale2  \n",
       "log-TPM                 \n",
       "b0002         9.783981  \n",
       "b0003         8.990193  \n",
       "b0004         9.040825  \n",
       "b0005         4.690402  \n",
       "b0006         7.552121  \n",
       "\n",
       "[5 rows x 105 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_cols = set(exp_df.columns.str[:-3])\n",
    "new_exp_df = pd.DataFrame()\n",
    "\n",
    "for i in unique_cols:\n",
    "    new_exp_df[i] = exp_df[[col for col in exp_df.columns if col[:-3]==i]].mean(axis=1)\n",
    "print(new_exp_df.shape)\n",
    "new_exp_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Double checking the work**\n",
    "\n",
    "Just making sure somethings add up here by taking an example column/condition and looking at the replicate values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ica__thm_gal__1</th>\n",
       "      <th>ica__thm_gal__2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log-TPM</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>11.092682</td>\n",
       "      <td>11.088257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>10.394027</td>\n",
       "      <td>10.411298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>10.601919</td>\n",
       "      <td>10.665099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>4.512912</td>\n",
       "      <td>4.859224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "      <td>6.723811</td>\n",
       "      <td>6.604000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         ica__thm_gal__1  ica__thm_gal__2\n",
       "log-TPM                                  \n",
       "b0002          11.092682        11.088257\n",
       "b0003          10.394027        10.411298\n",
       "b0004          10.601919        10.665099\n",
       "b0005           4.512912         4.859224\n",
       "b0006           6.723811         6.604000"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "example_col = list(unique_cols)[1]\n",
    "\n",
    "exp_df[[col for col in exp_df.columns if col[:-3]==example_col]].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And their average"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ica__thm_gal</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log-TPM</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>11.090470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>10.402663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>10.633509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>4.686068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "      <td>6.663905</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         ica__thm_gal\n",
       "log-TPM              \n",
       "b0002       11.090470\n",
       "b0003       10.402663\n",
       "b0004       10.633509\n",
       "b0005        4.686068\n",
       "b0006        6.663905"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_exp_df[[example_col]].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dealing with the weird duplicate column/s\n",
    "\n",
    "Time to kill any completely identical columns, these must be bugs on the data end and even though I could in theory keep one, their growth rate value is unclear (since it differs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Get an all-to-all correlation matrix between gene expression values\n",
    "temp_corr = new_exp_df.corr(method='spearman')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(105, 105)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pal__arab_ale14</th>\n",
       "      <th>ica__thm_gal</th>\n",
       "      <th>ssw__xyl_ale3</th>\n",
       "      <th>fur__delfur_fe2</th>\n",
       "      <th>ytf__wt_ph5</th>\n",
       "      <th>42c__42c_ale8</th>\n",
       "      <th>ytf__delyheO</th>\n",
       "      <th>cra_crp__delcra_glc</th>\n",
       "      <th>cra_crp__delcra_ac</th>\n",
       "      <th>pgi__pgi_ale1</th>\n",
       "      <th>...</th>\n",
       "      <th>rpoB__rpoBE546V_lb</th>\n",
       "      <th>ssw__wt_glyc</th>\n",
       "      <th>ica__gth</th>\n",
       "      <th>glu__glu_ale4</th>\n",
       "      <th>efeU__menFentCubiC_ale36</th>\n",
       "      <th>fps__fps_ptsI_ale1</th>\n",
       "      <th>42c__42c_ale2</th>\n",
       "      <th>omics__bw_ac</th>\n",
       "      <th>ytf__delyeiE</th>\n",
       "      <th>pgi__pgi_ale2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>pal__arab_ale14</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.880622</td>\n",
       "      <td>0.919062</td>\n",
       "      <td>0.915954</td>\n",
       "      <td>0.916830</td>\n",
       "      <td>0.867574</td>\n",
       "      <td>0.848017</td>\n",
       "      <td>0.922362</td>\n",
       "      <td>0.900065</td>\n",
       "      <td>0.918987</td>\n",
       "      <td>...</td>\n",
       "      <td>0.846204</td>\n",
       "      <td>0.845934</td>\n",
       "      <td>0.805012</td>\n",
       "      <td>0.891187</td>\n",
       "      <td>0.909152</td>\n",
       "      <td>0.924867</td>\n",
       "      <td>0.901711</td>\n",
       "      <td>0.931625</td>\n",
       "      <td>0.902226</td>\n",
       "      <td>0.875269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ica__thm_gal</th>\n",
       "      <td>0.880622</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.814639</td>\n",
       "      <td>0.845009</td>\n",
       "      <td>0.882751</td>\n",
       "      <td>0.801688</td>\n",
       "      <td>0.824915</td>\n",
       "      <td>0.888358</td>\n",
       "      <td>0.884646</td>\n",
       "      <td>0.846474</td>\n",
       "      <td>...</td>\n",
       "      <td>0.849325</td>\n",
       "      <td>0.801547</td>\n",
       "      <td>0.768087</td>\n",
       "      <td>0.882620</td>\n",
       "      <td>0.872295</td>\n",
       "      <td>0.898860</td>\n",
       "      <td>0.883579</td>\n",
       "      <td>0.903745</td>\n",
       "      <td>0.870752</td>\n",
       "      <td>0.860254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ssw__xyl_ale3</th>\n",
       "      <td>0.919062</td>\n",
       "      <td>0.814639</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.899916</td>\n",
       "      <td>0.901277</td>\n",
       "      <td>0.900105</td>\n",
       "      <td>0.857151</td>\n",
       "      <td>0.908824</td>\n",
       "      <td>0.851008</td>\n",
       "      <td>0.895229</td>\n",
       "      <td>...</td>\n",
       "      <td>0.779749</td>\n",
       "      <td>0.844000</td>\n",
       "      <td>0.796825</td>\n",
       "      <td>0.891669</td>\n",
       "      <td>0.900448</td>\n",
       "      <td>0.870549</td>\n",
       "      <td>0.902682</td>\n",
       "      <td>0.876793</td>\n",
       "      <td>0.895947</td>\n",
       "      <td>0.838098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fur__delfur_fe2</th>\n",
       "      <td>0.915954</td>\n",
       "      <td>0.845009</td>\n",
       "      <td>0.899916</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.951830</td>\n",
       "      <td>0.877977</td>\n",
       "      <td>0.844202</td>\n",
       "      <td>0.958872</td>\n",
       "      <td>0.924542</td>\n",
       "      <td>0.933114</td>\n",
       "      <td>...</td>\n",
       "      <td>0.807755</td>\n",
       "      <td>0.794648</td>\n",
       "      <td>0.869040</td>\n",
       "      <td>0.889556</td>\n",
       "      <td>0.942488</td>\n",
       "      <td>0.904895</td>\n",
       "      <td>0.900265</td>\n",
       "      <td>0.913742</td>\n",
       "      <td>0.904875</td>\n",
       "      <td>0.913807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ytf__wt_ph5</th>\n",
       "      <td>0.916830</td>\n",
       "      <td>0.882751</td>\n",
       "      <td>0.901277</td>\n",
       "      <td>0.951830</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.874055</td>\n",
       "      <td>0.891568</td>\n",
       "      <td>0.972327</td>\n",
       "      <td>0.915127</td>\n",
       "      <td>0.906419</td>\n",
       "      <td>...</td>\n",
       "      <td>0.836713</td>\n",
       "      <td>0.815491</td>\n",
       "      <td>0.874591</td>\n",
       "      <td>0.885708</td>\n",
       "      <td>0.964902</td>\n",
       "      <td>0.925843</td>\n",
       "      <td>0.892907</td>\n",
       "      <td>0.905586</td>\n",
       "      <td>0.956894</td>\n",
       "      <td>0.907995</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 105 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 pal__arab_ale14  ica__thm_gal  ssw__xyl_ale3  \\\n",
       "pal__arab_ale14              NaN      0.880622       0.919062   \n",
       "ica__thm_gal            0.880622           NaN       0.814639   \n",
       "ssw__xyl_ale3           0.919062      0.814639            NaN   \n",
       "fur__delfur_fe2         0.915954      0.845009       0.899916   \n",
       "ytf__wt_ph5             0.916830      0.882751       0.901277   \n",
       "\n",
       "                 fur__delfur_fe2  ytf__wt_ph5  42c__42c_ale8  ytf__delyheO  \\\n",
       "pal__arab_ale14         0.915954     0.916830       0.867574      0.848017   \n",
       "ica__thm_gal            0.845009     0.882751       0.801688      0.824915   \n",
       "ssw__xyl_ale3           0.899916     0.901277       0.900105      0.857151   \n",
       "fur__delfur_fe2              NaN     0.951830       0.877977      0.844202   \n",
       "ytf__wt_ph5             0.951830          NaN       0.874055      0.891568   \n",
       "\n",
       "                 cra_crp__delcra_glc  cra_crp__delcra_ac  pgi__pgi_ale1  ...  \\\n",
       "pal__arab_ale14             0.922362            0.900065       0.918987  ...   \n",
       "ica__thm_gal                0.888358            0.884646       0.846474  ...   \n",
       "ssw__xyl_ale3               0.908824            0.851008       0.895229  ...   \n",
       "fur__delfur_fe2             0.958872            0.924542       0.933114  ...   \n",
       "ytf__wt_ph5                 0.972327            0.915127       0.906419  ...   \n",
       "\n",
       "                 rpoB__rpoBE546V_lb  ssw__wt_glyc  ica__gth  glu__glu_ale4  \\\n",
       "pal__arab_ale14            0.846204      0.845934  0.805012       0.891187   \n",
       "ica__thm_gal               0.849325      0.801547  0.768087       0.882620   \n",
       "ssw__xyl_ale3              0.779749      0.844000  0.796825       0.891669   \n",
       "fur__delfur_fe2            0.807755      0.794648  0.869040       0.889556   \n",
       "ytf__wt_ph5                0.836713      0.815491  0.874591       0.885708   \n",
       "\n",
       "                 efeU__menFentCubiC_ale36  fps__fps_ptsI_ale1  42c__42c_ale2  \\\n",
       "pal__arab_ale14                  0.909152            0.924867       0.901711   \n",
       "ica__thm_gal                     0.872295            0.898860       0.883579   \n",
       "ssw__xyl_ale3                    0.900448            0.870549       0.902682   \n",
       "fur__delfur_fe2                  0.942488            0.904895       0.900265   \n",
       "ytf__wt_ph5                      0.964902            0.925843       0.892907   \n",
       "\n",
       "                 omics__bw_ac  ytf__delyeiE  pgi__pgi_ale2  \n",
       "pal__arab_ale14      0.931625      0.902226       0.875269  \n",
       "ica__thm_gal         0.903745      0.870752       0.860254  \n",
       "ssw__xyl_ale3        0.876793      0.895947       0.838098  \n",
       "fur__delfur_fe2      0.913742      0.904875       0.913807  \n",
       "ytf__wt_ph5          0.905586      0.956894       0.907995  \n",
       "\n",
       "[5 rows x 105 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###And make this into a symmetric dataframe\n",
    "temp_df = pd.DataFrame(\n",
    "    np.where(np.equal(*np.indices(temp_corr.shape)), np.nan, temp_corr.values),\n",
    "    temp_corr.index, temp_corr.columns\n",
    ")\n",
    "print(temp_df.shape)\n",
    "temp_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Identify columns that contain a value of \"1.\" since this indicates they have a perfect correlation**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pal__arab_ale14</th>\n",
       "      <th>ica__thm_gal</th>\n",
       "      <th>ssw__xyl_ale3</th>\n",
       "      <th>fur__delfur_fe2</th>\n",
       "      <th>ytf__wt_ph5</th>\n",
       "      <th>42c__42c_ale8</th>\n",
       "      <th>ytf__delyheO</th>\n",
       "      <th>cra_crp__delcra_glc</th>\n",
       "      <th>cra_crp__delcra_ac</th>\n",
       "      <th>pgi__pgi_ale1</th>\n",
       "      <th>...</th>\n",
       "      <th>rpoB__rpoBE546V_lb</th>\n",
       "      <th>ssw__wt_glyc</th>\n",
       "      <th>ica__gth</th>\n",
       "      <th>glu__glu_ale4</th>\n",
       "      <th>efeU__menFentCubiC_ale36</th>\n",
       "      <th>fps__fps_ptsI_ale1</th>\n",
       "      <th>42c__42c_ale2</th>\n",
       "      <th>omics__bw_ac</th>\n",
       "      <th>ytf__delyeiE</th>\n",
       "      <th>pgi__pgi_ale2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>pal__lyx_ale4</th>\n",
       "      <td>0.93004</td>\n",
       "      <td>0.883647</td>\n",
       "      <td>0.888263</td>\n",
       "      <td>0.878752</td>\n",
       "      <td>0.869489</td>\n",
       "      <td>0.828383</td>\n",
       "      <td>0.8044</td>\n",
       "      <td>0.88225</td>\n",
       "      <td>0.877665</td>\n",
       "      <td>0.898932</td>\n",
       "      <td>...</td>\n",
       "      <td>0.836452</td>\n",
       "      <td>0.815217</td>\n",
       "      <td>0.766783</td>\n",
       "      <td>0.882527</td>\n",
       "      <td>0.868086</td>\n",
       "      <td>0.917759</td>\n",
       "      <td>0.888757</td>\n",
       "      <td>0.924265</td>\n",
       "      <td>0.854234</td>\n",
       "      <td>0.852095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pal__lyx_ale2</th>\n",
       "      <td>0.93004</td>\n",
       "      <td>0.883647</td>\n",
       "      <td>0.888263</td>\n",
       "      <td>0.878752</td>\n",
       "      <td>0.869489</td>\n",
       "      <td>0.828383</td>\n",
       "      <td>0.8044</td>\n",
       "      <td>0.88225</td>\n",
       "      <td>0.877665</td>\n",
       "      <td>0.898932</td>\n",
       "      <td>...</td>\n",
       "      <td>0.836452</td>\n",
       "      <td>0.815217</td>\n",
       "      <td>0.766783</td>\n",
       "      <td>0.882527</td>\n",
       "      <td>0.868086</td>\n",
       "      <td>0.917759</td>\n",
       "      <td>0.888757</td>\n",
       "      <td>0.924265</td>\n",
       "      <td>0.854234</td>\n",
       "      <td>0.852095</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 105 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               pal__arab_ale14  ica__thm_gal  ssw__xyl_ale3  fur__delfur_fe2  \\\n",
       "pal__lyx_ale4          0.93004      0.883647       0.888263         0.878752   \n",
       "pal__lyx_ale2          0.93004      0.883647       0.888263         0.878752   \n",
       "\n",
       "               ytf__wt_ph5  42c__42c_ale8  ytf__delyheO  cra_crp__delcra_glc  \\\n",
       "pal__lyx_ale4     0.869489       0.828383        0.8044              0.88225   \n",
       "pal__lyx_ale2     0.869489       0.828383        0.8044              0.88225   \n",
       "\n",
       "               cra_crp__delcra_ac  pgi__pgi_ale1  ...  rpoB__rpoBE546V_lb  \\\n",
       "pal__lyx_ale4            0.877665       0.898932  ...            0.836452   \n",
       "pal__lyx_ale2            0.877665       0.898932  ...            0.836452   \n",
       "\n",
       "               ssw__wt_glyc  ica__gth  glu__glu_ale4  \\\n",
       "pal__lyx_ale4      0.815217  0.766783       0.882527   \n",
       "pal__lyx_ale2      0.815217  0.766783       0.882527   \n",
       "\n",
       "               efeU__menFentCubiC_ale36  fps__fps_ptsI_ale1  42c__42c_ale2  \\\n",
       "pal__lyx_ale4                  0.868086            0.917759       0.888757   \n",
       "pal__lyx_ale2                  0.868086            0.917759       0.888757   \n",
       "\n",
       "               omics__bw_ac  ytf__delyeiE  pgi__pgi_ale2  \n",
       "pal__lyx_ale4      0.924265      0.854234       0.852095  \n",
       "pal__lyx_ale2      0.924265      0.854234       0.852095  \n",
       "\n",
       "[2 rows x 105 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp_df[temp_df.values==1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get rid of them both since there is obviously an error here somewhere"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3923, 103)\n"
     ]
    }
   ],
   "source": [
    "new_exp_df.drop(['pal__lyx_ale2', 'pal__lyx_ale4'], axis=1, inplace=True)\n",
    "print(new_exp_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Average the growth rates across these replicates in the metadata as well"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(173, 27)\n",
      "(105, 25)\n",
      "(105, 26)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Study</th>\n",
       "      <th>Project ID</th>\n",
       "      <th>Condition ID</th>\n",
       "      <th>Strain Description</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Base Media</th>\n",
       "      <th>Carbon Source (g/L)</th>\n",
       "      <th>Nitrogen Source (g/L)</th>\n",
       "      <th>Electron Acceptor</th>\n",
       "      <th>...</th>\n",
       "      <th>Sequencing Machine</th>\n",
       "      <th>Additional Details</th>\n",
       "      <th>DOI</th>\n",
       "      <th>GEO</th>\n",
       "      <th>Simple_sample_id</th>\n",
       "      <th>Temperature (C)</th>\n",
       "      <th>pH</th>\n",
       "      <th>Growth Rate (1/hr)</th>\n",
       "      <th>Alignment</th>\n",
       "      <th>Replicate #</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>42c__42c_ale1__1</td>\n",
       "      <td>42C Evolution</td>\n",
       "      <td>42c</td>\n",
       "      <td>42c_ale1</td>\n",
       "      <td>Escherichia coli 42C.1.124.1</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>O2</td>\n",
       "      <td>...</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>42C A1 F124 I1</td>\n",
       "      <td>doi.org/10.1093/molbev/msu209</td>\n",
       "      <td>GSE132442</td>\n",
       "      <td>42c__42c_ale1</td>\n",
       "      <td>42.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.95</td>\n",
       "      <td>98.40</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>42c__42c_ale10__1</td>\n",
       "      <td>42C Evolution</td>\n",
       "      <td>42c</td>\n",
       "      <td>42c_ale10</td>\n",
       "      <td>Escherichia coli 42C.10.153.1</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>O2</td>\n",
       "      <td>...</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>42C A10 F153 I1</td>\n",
       "      <td>doi.org/10.1093/molbev/msu209</td>\n",
       "      <td>GSE132442</td>\n",
       "      <td>42c__42c_ale10</td>\n",
       "      <td>42.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.98</td>\n",
       "      <td>96.91</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>42c__42c_ale2__1</td>\n",
       "      <td>42C Evolution</td>\n",
       "      <td>42c</td>\n",
       "      <td>42c_ale2</td>\n",
       "      <td>Escherichia coli 42C.2.163.1</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>O2</td>\n",
       "      <td>...</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>42C A2 F163 I1</td>\n",
       "      <td>doi.org/10.1093/molbev/msu209</td>\n",
       "      <td>GSE132442</td>\n",
       "      <td>42c__42c_ale2</td>\n",
       "      <td>42.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.97</td>\n",
       "      <td>97.51</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>42c__42c_ale3__1</td>\n",
       "      <td>42C Evolution</td>\n",
       "      <td>42c</td>\n",
       "      <td>42c_ale3</td>\n",
       "      <td>Escherichia coli 42C.3.120.1</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>O2</td>\n",
       "      <td>...</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>42C A3 F120 I1</td>\n",
       "      <td>doi.org/10.1093/molbev/msu209</td>\n",
       "      <td>GSE132442</td>\n",
       "      <td>42c__42c_ale3</td>\n",
       "      <td>42.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.92</td>\n",
       "      <td>97.13</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>42c__42c_ale4__1</td>\n",
       "      <td>42C Evolution</td>\n",
       "      <td>42c</td>\n",
       "      <td>42c_ale4</td>\n",
       "      <td>Escherichia coli 42C.4.161.1</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>O2</td>\n",
       "      <td>...</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>42C A4 F161 I1</td>\n",
       "      <td>doi.org/10.1093/molbev/msu209</td>\n",
       "      <td>GSE132442</td>\n",
       "      <td>42c__42c_ale4</td>\n",
       "      <td>42.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.03</td>\n",
       "      <td>97.07</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Sample ID          Study Project ID Condition ID  \\\n",
       "132   42c__42c_ale1__1  42C Evolution        42c     42c_ale1   \n",
       "140  42c__42c_ale10__1  42C Evolution        42c    42c_ale10   \n",
       "133   42c__42c_ale2__1  42C Evolution        42c     42c_ale2   \n",
       "134   42c__42c_ale3__1  42C Evolution        42c     42c_ale3   \n",
       "135   42c__42c_ale4__1  42C Evolution        42c     42c_ale4   \n",
       "\n",
       "                Strain Description  Strain Base Media Carbon Source (g/L)  \\\n",
       "132   Escherichia coli 42C.1.124.1  MG1655         M9          glucose(4)   \n",
       "140  Escherichia coli 42C.10.153.1  MG1655         M9          glucose(4)   \n",
       "133   Escherichia coli 42C.2.163.1  MG1655         M9          glucose(4)   \n",
       "134   Escherichia coli 42C.3.120.1  MG1655         M9          glucose(4)   \n",
       "135   Escherichia coli 42C.4.161.1  MG1655         M9          glucose(4)   \n",
       "\n",
       "    Nitrogen Source (g/L) Electron Acceptor  ... Sequencing Machine  \\\n",
       "132              NH4Cl(1)                O2  ...              MiSeq   \n",
       "140              NH4Cl(1)                O2  ...              MiSeq   \n",
       "133              NH4Cl(1)                O2  ...              MiSeq   \n",
       "134              NH4Cl(1)                O2  ...              MiSeq   \n",
       "135              NH4Cl(1)                O2  ...              MiSeq   \n",
       "\n",
       "    Additional Details                            DOI        GEO  \\\n",
       "132     42C A1 F124 I1  doi.org/10.1093/molbev/msu209  GSE132442   \n",
       "140    42C A10 F153 I1  doi.org/10.1093/molbev/msu209  GSE132442   \n",
       "133     42C A2 F163 I1  doi.org/10.1093/molbev/msu209  GSE132442   \n",
       "134     42C A3 F120 I1  doi.org/10.1093/molbev/msu209  GSE132442   \n",
       "135     42C A4 F161 I1  doi.org/10.1093/molbev/msu209  GSE132442   \n",
       "\n",
       "    Simple_sample_id Temperature (C)   pH Growth Rate (1/hr) Alignment  \\\n",
       "132    42c__42c_ale1            42.0  7.0               0.95     98.40   \n",
       "140   42c__42c_ale10            42.0  7.0               0.98     96.91   \n",
       "133    42c__42c_ale2            42.0  7.0               0.97     97.51   \n",
       "134    42c__42c_ale3            42.0  7.0               0.92     97.13   \n",
       "135    42c__42c_ale4            42.0  7.0               1.03     97.07   \n",
       "\n",
       "    Replicate #  \n",
       "132           1  \n",
       "140           1  \n",
       "133           1  \n",
       "134           1  \n",
       "135           1  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###Assign a unique id that removes the replicate information\n",
    "meta_df['Simple_sample_id'] = meta_df['Sample ID'].str[:-3]\n",
    "print(meta_df.shape)\n",
    "###Group according to this new id\n",
    "group_cols = ['Simple_sample_id']\n",
    "###For these numeric columns I'll take the mean\n",
    "metric_cols_a = ['Temperature (C)', 'pH', 'Growth Rate (1/hr)', 'Alignment']\n",
    "aggs_a = meta_df.groupby(group_cols)[metric_cols_a].mean()\n",
    "###And for these I'll just grab the count\n",
    "metric_cols_b = ['Replicate #', 'Biological Replicates']\n",
    "aggs_b = meta_df.groupby(group_cols)['Replicate #'].count()\n",
    "\n",
    "###Drop the columns from the original dataframe (we'll add them back in later)\n",
    "meta_df.drop(metric_cols_a, axis=1, inplace=True)\n",
    "meta_df.drop(metric_cols_b, axis=1, inplace=True)\n",
    "###And duplicates\n",
    "meta_df.drop_duplicates(subset=group_cols, keep='first', inplace=True)\n",
    "\n",
    "###Now merge the main dataframe with the grouped ones\n",
    "meta_df = meta_df.merge(right=aggs_a, right_index=True, left_on=group_cols, how='right')\n",
    "print(meta_df.shape)\n",
    "meta_df = meta_df.merge(right=aggs_b, right_index=True, left_on=group_cols, how='right')\n",
    "print(meta_df.shape)\n",
    "meta_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**And get rid of those problematic samples from this dataframe as well**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(105, 26)\n",
      "(104, 26)\n",
      "(103, 26)\n"
     ]
    }
   ],
   "source": [
    "print(meta_df.shape)\n",
    "meta_df = meta_df[meta_df['Sample ID'].str.contains('pal__lyx_ale2__')==False]\n",
    "print(meta_df.shape)\n",
    "meta_df = meta_df[meta_df['Sample ID'].str.contains('pal__lyx_ale4__')==False]\n",
    "print(meta_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Sum should equal the original shape!**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2    61\n",
       "1    41\n",
       "6     1\n",
       "Name: Replicate #, dtype: int64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_df['Replicate #'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Make sure that the columns line up when matching across these two dataframes** "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_exp_df = new_exp_df[meta_df['Simple_sample_id']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert all(new_exp_df.columns == meta_df['Simple_sample_id'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**And add the doubling time just for good measure**\n",
    "\n",
    "Which is just a slight transformation of growth rate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_df['Doubling_time'] = np.log(2)/meta_df['Growth Rate (1/hr)']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Construct a third `dataframe` containing gene expression data summary stats\n",
    "\n",
    "Strictly speaking this isn't super necessary but might as well do it now to get it done and over with"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "exp_summary_df = new_exp_df.apply(pd.DataFrame.describe, axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**The % signs seem to cause some problems down the road so lets remove them**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['count', 'mean', 'std', 'min', '25_percentile', '50_percentile', '75_percentile', 'max']\n"
     ]
    }
   ],
   "source": [
    "col_listy = []\n",
    "for col in exp_summary_df.columns:\n",
    "    if '%' not in col:\n",
    "        col_listy.append(col)\n",
    "    else:\n",
    "        col_listy.append(col.replace('%', '_percentile'))\n",
    "print(col_listy)\n",
    "exp_summary_df.columns = col_listy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**And add some other dispersion metrics**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "exp_summary_df['cv'] = exp_summary_df['std']/exp_summary_df['mean']\n",
    "exp_summary_df['noise'] = exp_summary_df['std'].pow(2)/exp_summary_df['mean'].pow(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Finally, adding some of (what we think are) the cool new variables to consider**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "slopes = []\n",
    "pearsons = []\n",
    "spearmans = []\n",
    "for gene in exp_summary_df.index:\n",
    "    a, b, c, d, e = stats.linregress(new_exp_df.loc[gene], meta_df['Growth Rate (1/hr)'])\n",
    "    slopes.append(a)\n",
    "    pearsons.append(c)\n",
    "    rho, p = stats.spearmanr(new_exp_df.loc[gene], meta_df['Growth Rate (1/hr)'])\n",
    "    spearmans.append(rho)\n",
    "    \n",
    "exp_summary_df['lin_slope'] = slopes\n",
    "exp_summary_df['lin_r'] = pearsons\n",
    "exp_summary_df['spearmans_rho'] = spearmans"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Save some files\n",
    "\n",
    "This was the whole point of all the code above. Should have some straightforward data tables now"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3923, 103)\n",
      "(103, 27)\n",
      "(3923, 13)\n"
     ]
    }
   ],
   "source": [
    "print(new_exp_df.shape)\n",
    "print(meta_df.shape)\n",
    "print(exp_summary_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_exp_df.to_csv('../Data/processed_data/processed_expression_ecoli.tsv', sep='\\t')\n",
    "meta_df.to_csv('../Data/processed_data/processed_metadata_ecoli.tsv', sep='\\t')\n",
    "exp_summary_df.to_csv('../Data/processed_data/processed_expression_summary_ecoli.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get a thinned down dataset as a robustness check\n",
    "\n",
    "Since conditions are correlated with one another, I'm constructing a more sparsely populated dataset where inter-condition correlations are minimized using a greedy algorithm. \n",
    "\n",
    "The algorithm works as follows:\n",
    "1. Find which two conditions are the most highly correlated across the entire all-to-all correlation matrix\n",
    "2. Randomly delete one of the two conditions in question\n",
    "3. Iterate to (1)\n",
    "4. Stop after reaching a pre-defined final dataset size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "random.seed(42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>42c__42c_ale1</th>\n",
       "      <th>42c__42c_ale10</th>\n",
       "      <th>42c__42c_ale2</th>\n",
       "      <th>42c__42c_ale3</th>\n",
       "      <th>42c__42c_ale4</th>\n",
       "      <th>42c__42c_ale5</th>\n",
       "      <th>42c__42c_ale6</th>\n",
       "      <th>42c__42c_ale8</th>\n",
       "      <th>42c__42c_ale9</th>\n",
       "      <th>42c__wt_42c</th>\n",
       "      <th>...</th>\n",
       "      <th>ytf__delydcI_ph5</th>\n",
       "      <th>ytf__delydcI_ph8</th>\n",
       "      <th>ytf__delyddM</th>\n",
       "      <th>ytf__delyeiE</th>\n",
       "      <th>ytf__delyheO</th>\n",
       "      <th>ytf__delyiaJ</th>\n",
       "      <th>ytf__delyieP</th>\n",
       "      <th>ytf__wt_glc</th>\n",
       "      <th>ytf__wt_ph5</th>\n",
       "      <th>ytf__wt_ph8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>42c__42c_ale1</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.927120</td>\n",
       "      <td>0.882641</td>\n",
       "      <td>0.862337</td>\n",
       "      <td>0.890492</td>\n",
       "      <td>0.844707</td>\n",
       "      <td>0.934393</td>\n",
       "      <td>0.910083</td>\n",
       "      <td>0.853765</td>\n",
       "      <td>0.836816</td>\n",
       "      <td>...</td>\n",
       "      <td>0.806675</td>\n",
       "      <td>0.798028</td>\n",
       "      <td>0.828742</td>\n",
       "      <td>0.840311</td>\n",
       "      <td>0.902051</td>\n",
       "      <td>0.864096</td>\n",
       "      <td>0.889329</td>\n",
       "      <td>0.841736</td>\n",
       "      <td>0.803002</td>\n",
       "      <td>0.802468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42c__42c_ale10</th>\n",
       "      <td>0.927120</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.931647</td>\n",
       "      <td>0.909807</td>\n",
       "      <td>0.963706</td>\n",
       "      <td>0.902406</td>\n",
       "      <td>0.919654</td>\n",
       "      <td>0.961923</td>\n",
       "      <td>0.890414</td>\n",
       "      <td>0.843208</td>\n",
       "      <td>...</td>\n",
       "      <td>0.836582</td>\n",
       "      <td>0.841167</td>\n",
       "      <td>0.846182</td>\n",
       "      <td>0.830256</td>\n",
       "      <td>0.861610</td>\n",
       "      <td>0.841320</td>\n",
       "      <td>0.898360</td>\n",
       "      <td>0.876208</td>\n",
       "      <td>0.832988</td>\n",
       "      <td>0.839818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42c__42c_ale2</th>\n",
       "      <td>0.882641</td>\n",
       "      <td>0.931647</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.964109</td>\n",
       "      <td>0.956908</td>\n",
       "      <td>0.968556</td>\n",
       "      <td>0.856027</td>\n",
       "      <td>0.918062</td>\n",
       "      <td>0.963966</td>\n",
       "      <td>0.902696</td>\n",
       "      <td>...</td>\n",
       "      <td>0.899551</td>\n",
       "      <td>0.901670</td>\n",
       "      <td>0.888363</td>\n",
       "      <td>0.876668</td>\n",
       "      <td>0.827253</td>\n",
       "      <td>0.818446</td>\n",
       "      <td>0.865943</td>\n",
       "      <td>0.904547</td>\n",
       "      <td>0.892907</td>\n",
       "      <td>0.902017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42c__42c_ale3</th>\n",
       "      <td>0.862337</td>\n",
       "      <td>0.909807</td>\n",
       "      <td>0.964109</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.938152</td>\n",
       "      <td>0.974434</td>\n",
       "      <td>0.858418</td>\n",
       "      <td>0.896590</td>\n",
       "      <td>0.965164</td>\n",
       "      <td>0.900097</td>\n",
       "      <td>...</td>\n",
       "      <td>0.897568</td>\n",
       "      <td>0.896120</td>\n",
       "      <td>0.888025</td>\n",
       "      <td>0.880415</td>\n",
       "      <td>0.811377</td>\n",
       "      <td>0.803547</td>\n",
       "      <td>0.851940</td>\n",
       "      <td>0.880488</td>\n",
       "      <td>0.888256</td>\n",
       "      <td>0.896066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42c__42c_ale4</th>\n",
       "      <td>0.890492</td>\n",
       "      <td>0.963706</td>\n",
       "      <td>0.956908</td>\n",
       "      <td>0.938152</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.930396</td>\n",
       "      <td>0.887814</td>\n",
       "      <td>0.955125</td>\n",
       "      <td>0.919252</td>\n",
       "      <td>0.875380</td>\n",
       "      <td>...</td>\n",
       "      <td>0.869244</td>\n",
       "      <td>0.874356</td>\n",
       "      <td>0.863983</td>\n",
       "      <td>0.849296</td>\n",
       "      <td>0.835690</td>\n",
       "      <td>0.823237</td>\n",
       "      <td>0.879365</td>\n",
       "      <td>0.894170</td>\n",
       "      <td>0.863277</td>\n",
       "      <td>0.872864</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 103 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                42c__42c_ale1  42c__42c_ale10  42c__42c_ale2  42c__42c_ale3  \\\n",
       "42c__42c_ale1        1.000000        0.927120       0.882641       0.862337   \n",
       "42c__42c_ale10       0.927120        1.000000       0.931647       0.909807   \n",
       "42c__42c_ale2        0.882641        0.931647       1.000000       0.964109   \n",
       "42c__42c_ale3        0.862337        0.909807       0.964109       1.000000   \n",
       "42c__42c_ale4        0.890492        0.963706       0.956908       0.938152   \n",
       "\n",
       "                42c__42c_ale4  42c__42c_ale5  42c__42c_ale6  42c__42c_ale8  \\\n",
       "42c__42c_ale1        0.890492       0.844707       0.934393       0.910083   \n",
       "42c__42c_ale10       0.963706       0.902406       0.919654       0.961923   \n",
       "42c__42c_ale2        0.956908       0.968556       0.856027       0.918062   \n",
       "42c__42c_ale3        0.938152       0.974434       0.858418       0.896590   \n",
       "42c__42c_ale4        1.000000       0.930396       0.887814       0.955125   \n",
       "\n",
       "                42c__42c_ale9  42c__wt_42c  ...  ytf__delydcI_ph5  \\\n",
       "42c__42c_ale1        0.853765     0.836816  ...          0.806675   \n",
       "42c__42c_ale10       0.890414     0.843208  ...          0.836582   \n",
       "42c__42c_ale2        0.963966     0.902696  ...          0.899551   \n",
       "42c__42c_ale3        0.965164     0.900097  ...          0.897568   \n",
       "42c__42c_ale4        0.919252     0.875380  ...          0.869244   \n",
       "\n",
       "                ytf__delydcI_ph8  ytf__delyddM  ytf__delyeiE  ytf__delyheO  \\\n",
       "42c__42c_ale1           0.798028      0.828742      0.840311      0.902051   \n",
       "42c__42c_ale10          0.841167      0.846182      0.830256      0.861610   \n",
       "42c__42c_ale2           0.901670      0.888363      0.876668      0.827253   \n",
       "42c__42c_ale3           0.896120      0.888025      0.880415      0.811377   \n",
       "42c__42c_ale4           0.874356      0.863983      0.849296      0.835690   \n",
       "\n",
       "                ytf__delyiaJ  ytf__delyieP  ytf__wt_glc  ytf__wt_ph5  \\\n",
       "42c__42c_ale1       0.864096      0.889329     0.841736     0.803002   \n",
       "42c__42c_ale10      0.841320      0.898360     0.876208     0.832988   \n",
       "42c__42c_ale2       0.818446      0.865943     0.904547     0.892907   \n",
       "42c__42c_ale3       0.803547      0.851940     0.880488     0.888256   \n",
       "42c__42c_ale4       0.823237      0.879365     0.894170     0.863277   \n",
       "\n",
       "                ytf__wt_ph8  \n",
       "42c__42c_ale1      0.802468  \n",
       "42c__42c_ale10     0.839818  \n",
       "42c__42c_ale2      0.902017  \n",
       "42c__42c_ale3      0.896066  \n",
       "42c__42c_ale4      0.872864  \n",
       "\n",
       "[5 rows x 103 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cond_corr_mat = new_exp_df.corr(method='spearman')\n",
    "cond_corr_mat.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**The `final_data_size` is of course completely arbitrary. But the point is to get rid of some correlated data so it does the trick.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_data_size = 30\n",
    "n_to_prune = cond_corr_mat.shape[1] - final_data_size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.fill_diagonal(cond_corr_mat.values, np.nan)\n",
    "to_prune = []\n",
    "for i in range(n_to_prune):\n",
    "    tempy = cond_corr_mat.loc[[i for i in cond_corr_mat.columns if i not in to_prune]][[i for i in cond_corr_mat.columns if i not in to_prune]].max()\n",
    "    to_prune.append(random.choice([tempy.sort_values(ascending=False).index[0],\\\n",
    "                                   tempy.sort_values(ascending=False).index[1]]))\n",
    "np.fill_diagonal(cond_corr_mat.values, 1.)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Remove relevant rows from two of the dataframes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "to_keep = [i for i in cond_corr_mat.columns if i not in to_prune]\n",
    "new_exp_sprs_df = new_exp_df[to_keep]\n",
    "meta_sprs_df = meta_df[meta_df['Simple_sample_id'].isin(to_keep)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**And regenerate the summary dataframe to only use this thinned-down set of samples**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['count', 'mean', 'std', 'min', '25_percentile', '50_percentile', '75_percentile', 'max']\n"
     ]
    }
   ],
   "source": [
    "exp_summary_sprs_df = new_exp_sprs_df.apply(pd.DataFrame.describe, axis=1)\n",
    "col_listy = []\n",
    "for col in exp_summary_sprs_df.columns:\n",
    "    if '%' not in col:\n",
    "        col_listy.append(col)\n",
    "    else:\n",
    "        col_listy.append(col.replace('%', '_percentile'))\n",
    "print(col_listy)\n",
    "exp_summary_sprs_df.columns = col_listy\n",
    "\n",
    "exp_summary_sprs_df['cv'] = exp_summary_sprs_df['std']/exp_summary_sprs_df['mean']\n",
    "exp_summary_sprs_df['noise'] = exp_summary_sprs_df['std'].pow(2)/exp_summary_sprs_df['mean'].pow(2)\n",
    "\n",
    "slopes = []\n",
    "pearsons = []\n",
    "spearmans = []\n",
    "for gene in exp_summary_sprs_df.index:\n",
    "    a, b, c, d, e = stats.linregress(new_exp_sprs_df.loc[gene], meta_sprs_df['Growth Rate (1/hr)'])\n",
    "    slopes.append(a)\n",
    "    pearsons.append(c)\n",
    "    rho, p = stats.spearmanr(new_exp_sprs_df.loc[gene], meta_sprs_df['Growth Rate (1/hr)'])\n",
    "    spearmans.append(rho)\n",
    "    \n",
    "exp_summary_sprs_df['lin_slope'] = slopes\n",
    "exp_summary_sprs_df['lin_r'] = pearsons\n",
    "exp_summary_sprs_df['spearmans_rho'] = spearmans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3923, 30)\n",
      "(30, 27)\n",
      "(3923, 13)\n"
     ]
    }
   ],
   "source": [
    "print(new_exp_sprs_df.shape)\n",
    "print(meta_sprs_df.shape)\n",
    "print(exp_summary_sprs_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_exp_sprs_df.to_csv('../Data/processed_data/processed_expression_ecoli_SPARSE.tsv', sep='\\t')\n",
    "meta_sprs_df.to_csv('../Data/processed_data/processed_metadata_ecoli_SPARSE.tsv', sep='\\t')\n",
    "exp_summary_sprs_df.to_csv('../Data/processed_data/processed_expression_summary_ecoli_SPARSE.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create neutral data sets for additional robustness check\n",
    "**Limit analysis to non-ALE (adaptive laboratory environment) strains to control for selection**\n",
    "**First, remove ALE strains in data sets**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(48, 27)\n"
     ]
    }
   ],
   "source": [
    "meta_ntrl_df = meta_df[meta_df['Evolved Sample'].str.contains('No')==True]\n",
    "print(meta_ntrl_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3923, 48)\n"
     ]
    }
   ],
   "source": [
    "new_exp_ntrl_df = new_exp_df[meta_ntrl_df['Simple_sample_id']]\n",
    "print(new_exp_ntrl_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**And regenerate the summary dataframe to only use this non-ALE set of samples**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['count', 'mean', 'std', 'min', '25_percentile', '50_percentile', '75_percentile', 'max']\n"
     ]
    }
   ],
   "source": [
    "exp_summary_ntrl_df = new_exp_ntrl_df.apply(pd.DataFrame.describe, axis=1)\n",
    "col_listy = []\n",
    "for col in exp_summary_ntrl_df.columns:\n",
    "    if '%' not in col:\n",
    "        col_listy.append(col)\n",
    "    else:\n",
    "        col_listy.append(col.replace('%', '_percentile'))\n",
    "print(col_listy)\n",
    "exp_summary_ntrl_df.columns = col_listy\n",
    "\n",
    "exp_summary_ntrl_df['cv'] = exp_summary_ntrl_df['std']/exp_summary_ntrl_df['mean']\n",
    "exp_summary_ntrl_df['noise'] = exp_summary_ntrl_df['std'].pow(2)/exp_summary_ntrl_df['mean'].pow(2)\n",
    "\n",
    "slopes = []\n",
    "pearsons = []\n",
    "spearmans = []\n",
    "for gene in exp_summary_ntrl_df.index:\n",
    "    a, b, c, d, e = stats.linregress(new_exp_ntrl_df.loc[gene], meta_ntrl_df['Growth Rate (1/hr)'])\n",
    "    slopes.append(a)\n",
    "    pearsons.append(c)\n",
    "    rho, p = stats.spearmanr(new_exp_ntrl_df.loc[gene], meta_ntrl_df['Growth Rate (1/hr)'])\n",
    "    spearmans.append(rho)\n",
    "    \n",
    "exp_summary_ntrl_df['lin_slope'] = slopes\n",
    "exp_summary_ntrl_df['lin_r'] = pearsons\n",
    "exp_summary_ntrl_df['spearmans_rho'] = spearmans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3923, 48)\n",
      "(48, 27)\n",
      "(3923, 13)\n"
     ]
    }
   ],
   "source": [
    "print(new_exp_ntrl_df.shape)\n",
    "print(meta_ntrl_df.shape)\n",
    "print(exp_summary_ntrl_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Save the files for use later in the pipeline**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_exp_ntrl_df.to_csv('../Data/processed_data/processed_expression_ecoli_NEUTRAL.tsv', sep='\\t')\n",
    "meta_ntrl_df.to_csv('../Data/processed_data/processed_metadata_ecoli_NEUTRAL.tsv', sep='\\t')\n",
    "exp_summary_ntrl_df.to_csv('../Data/processed_data/processed_expression_summary_ecoli_NEUTRAL.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Limit analysis to non-ALE, non-mutants, and non-knock-outs**\n",
    "\n",
    "**First, filter out mutants and knock-outs from neutral data sets**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(32, 27)\n"
     ]
    }
   ],
   "source": [
    "# remove knock-outs\n",
    "temp_meta_ntrl_wt_df = meta_ntrl_df[meta_ntrl_df['Strain Description'].str.contains('del')==False]\n",
    "print(temp_meta_ntrl_wt_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(28, 27)\n"
     ]
    }
   ],
   "source": [
    "# remove mutants\n",
    "non_mut = ['Escherichia coli K-12 MG1655', 'Escherichia coli BW25113']\n",
    "meta_ntrl_wt_df = temp_meta_ntrl_wt_df[temp_meta_ntrl_wt_df['Strain Description'].isin(non_mut)]\n",
    "print(meta_ntrl_wt_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3923, 28)\n"
     ]
    }
   ],
   "source": [
    "# remove knock-outs and mutants\n",
    "new_exp_ntrl_wt_df = new_exp_df[meta_ntrl_wt_df['Simple_sample_id']]\n",
    "print(new_exp_ntrl_wt_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**And regenerate summary stats for WT neutral data set**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['count', 'mean', 'std', 'min', '25_percentile', '50_percentile', '75_percentile', 'max']\n"
     ]
    }
   ],
   "source": [
    "exp_summary_ntrl_wt_df = new_exp_ntrl_wt_df.apply(pd.DataFrame.describe, axis=1)\n",
    "col_listy = []\n",
    "for col in exp_summary_ntrl_wt_df.columns:\n",
    "    if '%' not in col:\n",
    "        col_listy.append(col)\n",
    "    else:\n",
    "        col_listy.append(col.replace('%', '_percentile'))\n",
    "print(col_listy)\n",
    "exp_summary_ntrl_wt_df.columns = col_listy\n",
    "\n",
    "exp_summary_ntrl_wt_df['cv'] = exp_summary_ntrl_wt_df['std']/exp_summary_ntrl_wt_df['mean']\n",
    "exp_summary_ntrl_wt_df['noise'] = exp_summary_ntrl_wt_df['std'].pow(2)/exp_summary_ntrl_wt_df['mean'].pow(2)\n",
    "\n",
    "slopes = []\n",
    "pearsons = []\n",
    "spearmans = []\n",
    "for gene in exp_summary_ntrl_wt_df.index:\n",
    "    a, b, c, d, e = stats.linregress(new_exp_ntrl_wt_df.loc[gene], meta_ntrl_wt_df['Growth Rate (1/hr)'])\n",
    "    slopes.append(a)\n",
    "    pearsons.append(c)\n",
    "    rho, p = stats.spearmanr(new_exp_ntrl_wt_df.loc[gene], meta_ntrl_wt_df['Growth Rate (1/hr)'])\n",
    "    spearmans.append(rho)\n",
    "    \n",
    "exp_summary_ntrl_wt_df['lin_slope'] = slopes\n",
    "exp_summary_ntrl_wt_df['lin_r'] = pearsons\n",
    "exp_summary_ntrl_wt_df['spearmans_rho'] = spearmans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3923, 28)\n",
      "(28, 27)\n",
      "(3923, 13)\n"
     ]
    }
   ],
   "source": [
    "print(new_exp_ntrl_wt_df.shape)\n",
    "print(meta_ntrl_wt_df.shape)\n",
    "print(exp_summary_ntrl_wt_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Save files for use later in pipeline**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_exp_ntrl_wt_df.to_csv('../Data/processed_data/processed_expression_ecoli_NEUTRAL_WT.tsv', sep='\\t')\n",
    "meta_ntrl_wt_df.to_csv('../Data/processed_data/processed_metadata_ecoli_NEUTRAL_WT.tsv', sep='\\t')\n",
    "exp_summary_ntrl_wt_df.to_csv('../Data/processed_data/processed_expression_summary_ecoli_NEUTRAL_WT.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  },
  "toc": {
   "colors": {
    "hover_highlight": "#DAA520",
    "navigate_num": "#000000",
    "navigate_text": "#333333",
    "running_highlight": "#FF0000",
    "selected_highlight": "#FFD700",
    "sidebar_border": "#EEEEEE",
    "wrapper_background": "#FFFFFF"
   },
   "moveMenuLeft": true,
   "nav_menu": {
    "height": "192px",
    "width": "252px"
   },
   "navigate_menu": true,
   "number_sections": true,
   "sideBar": true,
   "threshold": 4,
   "toc_cell": false,
   "toc_section_display": "block",
   "toc_window_display": false,
   "widenNotebook": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
