
---
title: "Exploring codon usage bias and predictors"
author: "Mackenzie Johnson"
date: "12/2/2021"
output: html_document
---

This document is the R version of `5-explore_cub_predictors.ipynb`, created for easier extension of the analysis and figures created by Adam.


```{r, message=FALSE}

library(tidyverse)
library(glue)
library(viridis)
library(cowplot)
library(ggtext)
library(broom)
library(ggdist)
library(distributional)
library(patchwork)
library(here)

```


## Defining constants and save directories

```{r, message=FALSE}

data_key <- tibble(
  set = c(
    'ecoli_full', 
    'ecoli_sparse', 
    'ecoli_neutral',
    'ecoli_neutral_wt',
    'ecoli_non_at_rich',
    'ecoli_prot',
    'scer_rna', 
    'scer_prot'
  ),
  meta_file = c(
    here("Data", "processed_data", "processed_metadata_ecoli.tsv"),
    here("Data", "processed_data", "processed_metadata_ecoli_SPARSE.tsv"),
    here("Data", "processed_data", "processed_metadata_ecoli_NEUTRAL.tsv"),
    here("Data", "processed_data", "processed_metadata_ecoli_NEUTRAL_WT.tsv"),
    here("Data", "processed_data", "processed_metadata_ecoli.tsv"),
    here("Data", "processed_data", "processed_metadata_ecoli_PROTEIN.tsv"),
    here("Data", "processed_data", "processed_metadata_scer_RNA.tsv"),
    here("Data", "processed_data", "processed_metadata_scer_PROTEIN.tsv")
  ),
  expression_file = c(
    here("Data", "processed_data", "processed_expression_ecoli.tsv"),
    here("Data", "processed_data", "processed_expression_ecoli_SPARSE.tsv"),
    here("Data", "processed_data", "processed_expression_ecoli_NEUTRAL.tsv"),
    here("Data", "processed_data", "processed_expression_ecoli_NEUTRAL_WT.tsv"),
    here("Data", "processed_data", "processed_expression_ecoli.tsv"),
    here("Data", "processed_data", "processed_expression_ecoli_PROTEIN.tsv"),
    here("Data", "processed_data", "processed_expression_scer_RNA.tsv"),
    here("Data", "processed_data", "processed_expression_scer_PROTEIN.tsv")
  ),
  summary_file = c(
    here("Data", "processed_data", "processed_expression_summary_ecoli.tsv"),
    here("Data", "processed_data", "processed_expression_summary_ecoli_SPARSE.tsv"),
    here("Data", "processed_data", "processed_expression_summary_ecoli_NEUTRAL.tsv"),
    here("Data", "processed_data", "processed_expression_summary_ecoli_NEUTRAL_WT.tsv"),
    here("Data", "processed_data", "processed_expression_summary_ecoli.tsv"),
    here("Data", "processed_data", "processed_expression_summary_ecoli_PROTEIN.tsv"),
    here("Data", "processed_data", "processed_expression_summary_scer_RNA.tsv"),
    here("Data", "processed_data", "processed_expression_summary_scer_PROTEIN.tsv")
  ),
  cub_file = c(
    here("Data", "ecoli_info", "current_ecoli_master_table.tsv"),
    here("Data", "ecoli_info", "current_ecoli_master_table.tsv"),
    here("Data", "ecoli_info", "current_ecoli_master_table.tsv"),
    here("Data", "ecoli_info", "current_ecoli_master_table.tsv"),
    here("Data", "ecoli_info", "current_ecoli_master_table.tsv"),
    here("Data", "ecoli_info", "current_ecoli_master_table.tsv"),
    here("Data", "scer_info", "current_scer_master_table.tsv"),
    here("Data", "scer_info", "current_scer_master_table.tsv")
  )
)

# select dataset for analysis
data_tag <- "ecoli_full" 
metric <- "CAI" #"CAI" 

```


## Data import

```{r, message=FALSE}

# get relevant file names for data set of interest
data_key %>% 
  filter(set == data_tag) -> analysis_files

# read in data
# gene-specific measurements
full_df <- read_tsv(analysis_files$cub_file)
full_df <- full_df %>% 
  drop_na(CAI) # replace with metric
full_df$roc_semppr_mean <- log10(full_df$roc_semppr_mean)

# meta-data related to experiments / growth conditions
meta_df <- read_tsv(analysis_files$meta_file)

# condition-specific expression information
exp_df <- read_tsv(analysis_files$expression_file)

# expression sumary statistics calculated from the condition-specific expression matrices
exp_sum_df <- read_tsv(analysis_files$summary_file)


```

```{r}

if (data_tag == "ecoli_non_at_rich") {
  
  # read in list of AT-rich genes from dos Reis et al. 2003
  at_rich_df <- read_csv(
    here('Data', 'ecoli_info', 'at_rich_genes.csv'),
    col_names = c('locus_tag', 'gene'),
    show_col_types = FALSE
  )
  
  full_df <- full_df %>% 
    mutate(at_rich = 
             case_when(
               locus_tag %in% at_rich_df$gene ~ TRUE,
               gene %in% at_rich_df$locus_tag ~ TRUE,
               TRUE ~ FALSE
             ))
  
  non_at_genes <- full_df %>% filter(at_rich == FALSE) %>% pull(locus_tag)
  
  # filter out AT-rich genes
  exp_df <- exp_df %>% 
    filter(`log-TPM` %in% non_at_genes)
  
  exp_sum_df <- exp_sum_df %>% 
    filter(`log-TPM` %in% non_at_genes)
  
}

```


#### Merge data

```{r, message=FALSE}

if (data_tag %in% c("ecoli_full", "ecoli_sparse", "ecoli_neutral", "ecoli_neutral_wt", "ecoli_non_at_rich")) {
  full_df <- inner_join(
    full_df, 
    exp_sum_df, 
    by = c("locus_tag" = "log-TPM")
  )
  rm(exp_sum_df)
} else if (data_tag == "ecoli_prot") {
  full_df <- inner_join(
    full_df, 
    exp_sum_df, 
    by = c("gene" = "Gene")
  )
  rm(exp_sum_df)
} else {
  full_df <- full_df %>% 
    filter(gene != "hypothetical protein") %>% 
    mutate(gene = str_sub(.$gene, start = 1, end = -2)) 
  full_df$gene <- toupper(full_df$gene)
  full_df<- inner_join(
    full_df, 
    exp_sum_df, 
    by = c("gene" = "Gene")
  )
  rm(exp_sum_df)
}

```


## Find the individual conditions whose gene expression is most (and least) correlated with CUB

```{r, message=FALSE}

if (data_tag %in% c("ecoli_full", "ecoli_non_at_rich")) {
  # get df
  corr_df <- full_df %>% 
    select(locus_tag, CAI) %>% 
    left_join(., exp_df, by = c("locus_tag" = "log-TPM")) 
  # correlations
  exp_cub_corr <- cor(corr_df$CAI, corr_df[3:105], method = "pearson") %>% 
    as_tibble() %>% 
    gather(
      .,
      key = "condition",
      value = "pearsons"
    ) %>% 
    mutate(r2 = pearsons^2)
} else if (data_tag == "ecoli_sparse") {
  # get df
  corr_df <- full_df %>% 
    select(locus_tag, CAI) %>% 
    left_join(., exp_df, by = c("locus_tag" = "log-TPM")) 
  # correlations
  exp_cub_corr <- cor(corr_df$CAI, corr_df[3:32], method = "pearson") %>% 
    as_tibble() %>% 
    gather(
      .,
      key = "condition",
      value = "pearsons"
    ) %>% 
    mutate(r2 = pearsons^2)
} else if (data_tag == "ecoli_neutral") {
  # get df
  corr_df <- full_df %>% 
    select(locus_tag, CAI) %>% 
    left_join(., exp_df, by = c("locus_tag" = "log-TPM")) 
  # correlations
  exp_cub_corr <- cor(corr_df$CAI, corr_df[3:50], method = "pearson") %>% 
    as_tibble() %>% 
    gather(
      .,
      key = "condition",
      value = "pearsons"
    ) %>% 
    mutate(r2 = pearsons^2)
} else if (data_tag == "ecoli_neutral_wt") {
  # get df
  corr_df <- full_df %>% 
    select(locus_tag, CAI) %>% 
    left_join(., exp_df, by = c("locus_tag" = "log-TPM")) 
  # correlations
  exp_cub_corr <- cor(corr_df$CAI, corr_df[3:30], method = "pearson") %>% 
    as_tibble() %>% 
    gather(
      .,
      key = "condition",
      value = "pearsons"
    ) %>% 
    mutate(r2 = pearsons^2)
} else if (data_tag == "ecoli_prot") {
  # get df
  corr_df <- full_df %>% 
    select(gene, CAI) %>% 
    left_join(., exp_df, by = c("gene" = "Gene")) 
   # correlations
  exp_cub_corr <- cor(corr_df$CAI, corr_df[3:22], method = "pearson") %>% 
    as_tibble() %>% 
    gather(
      .,
      key = "condition",
      value = "pearsons"
    ) %>% 
    mutate(r2 = pearsons^2)
} else {
  # get df
  corr_df <- full_df %>% 
    select(gene, CAI) %>% 
    left_join(., exp_df, by = c("gene" = "Gene")) 
  # correlations
  exp_cub_corr <- cor(corr_df$CAI, corr_df[3:16], method = "pearson") %>% 
    as_tibble() %>% 
    gather(
      .,
      key = "condition",
      value = "pearsons"
    ) %>% 
    mutate(r2 = pearsons^2)
}

min_corr <- exp_cub_corr %>% 
  filter(r2 == min(r2))
max_corr <- exp_cub_corr %>% 
  filter(r2 == max(r2))

if (data_tag %in% c("ecoli_full", "ecoli_sparse", "ecoli_neutral", "ecoli_non_at_rich")) {
  
  fig_corr_min <- corr_df %>% 
    ggplot() +
      aes(x = `ica__cytd_rib`, y = CAI) +
      geom_point(alpha = 0.5) +
      scale_x_continuous(name = "Condition expression (log TPM)") +
      theme_minimal_grid() +
      labs(title = glue('*R*<sup>2</sup> = {round(min_corr$r2, 3)}')) +
      theme(plot.title = element_markdown())
  
  fig_corr_max <- corr_df %>% 
    ggplot() +
      aes(x = `rpoB__rpoBE672K_lb`, y = CAI) +
      geom_point(alpha = 0.5) +
      scale_x_continuous(name = "Condition expression (log TPM)") +
      theme_minimal_grid() +
      labs(title = glue('*R*<sup>2</sup> = {round(max_corr$r2, 3)}')) +
      theme(plot.title = element_markdown())
  
} else if (data_tag == "ecoli_neutral_wt") {
  
  fig_corr_min <- corr_df %>% 
    ggplot() +
      aes(x = `ica__cytd_rib`, y = CAI) +
      geom_point(alpha = 0.5) +
      scale_x_continuous(name = "Condition expression") +
      theme_minimal_grid() +
      labs(title = glue('*R*<sup>2</sup> = {round(min_corr$r2, 3)}')) +
      theme(plot.title = element_markdown())
  
  fig_corr_max <- corr_df %>% 
    ggplot() +
      aes(x = `ica__arg_sbt`, y = CAI) +
      geom_point(alpha = 0.5) +
      scale_x_continuous(name = "Condition expression") +
      theme_minimal_grid() +
      labs(title = glue('*R*<sup>2</sup> = {round(max_corr$r2, 3)}')) +
      theme(plot.title = element_markdown())
  
} else if (data_tag == "ecoli_prot") {
  
  fig_corr_min <- corr_df %>% 
    ggplot() +
      aes(x = `Galactose`, y = CAI) +
      geom_point(alpha = 0.5) +
      scale_x_continuous(name = "Condition expression") +
      theme_minimal_grid() +
      labs(title = glue('*R*<sup>2</sup> = {round(min_corr$r2, 3)}')) +
      theme(plot.title = element_markdown())
  
  fig_corr_max <- corr_df %>% 
    ggplot() +
      aes(x = `LB`, y = CAI) +
      geom_point(alpha = 0.5) +
      scale_x_continuous(name = "Condition expression") +
      theme_minimal_grid() +
      labs(title = glue('*R*<sup>2</sup> = {round(max_corr$r2, 3)}')) +
      theme(plot.title = element_markdown())
  
} else if (data_tag == "scer_rna") {
  
  fig_corr_min <- corr_df %>% 
    ggplot() +
      aes(x = `NH4_0.05_7.5_0.5`, y = CAI) +
      geom_point(alpha = 0.5) +
      scale_x_continuous(name = "Condition expression") +
      theme_minimal_grid() +
      labs(title = glue('*R*<sup>2</sup> = {round(min_corr$r2, 3)}')) +
      theme(plot.title = element_markdown())
  
  fig_corr_max <- corr_df %>% 
    ggplot() +
      aes(x = `NH4_0.35_7.5_0.5`, y = CAI) +
      geom_point(alpha = 0.5) +
      scale_x_continuous(name = "Condition expression") +
      theme_minimal_grid() +
      labs(title = glue('*R*<sup>2</sup> = {round(max_corr$r2, 3)}')) +
      theme(plot.title = element_markdown())
  
} else {
  
   fig_corr_min <- corr_df %>% 
    ggplot() +
      aes(x = `Phe_0.1_7.5_12.35`, y = CAI) +
      geom_point(alpha = 0.5) +
      scale_x_continuous(name = "Condition expression") +
      theme_minimal_grid() +
      labs(title = glue('*R*<sup>2</sup> = {round(min_corr$r2, 3)}')) +
      theme(plot.title = element_markdown())
  
  fig_corr_max <- corr_df %>% 
    ggplot() +
      aes(x = `NH4_0.35_7.5_0.5`, y = CAI) +
      geom_point(alpha = 0.5) +
      scale_x_continuous(name = "Condition expression") +
      theme_minimal_grid() +
      labs(title = glue('*R*<sup>2</sup> = {round(max_corr$r2, 3)}')) +
      theme(plot.title = element_markdown())
  
}

fig_corr_grow <- meta_df %>% 
  select(Simple_sample_id, `Growth Rate (1/hr)`) %>% 
  inner_join(., exp_cub_corr, by = c("Simple_sample_id" = "condition")) %>% 
  ggplot() +
    aes(x = `Growth Rate (1/hr)`, y = r2) +
    geom_point(size = 2, alpha = 0.75) +
    scale_y_continuous(name = '*R*<sup>2</sup> (CAI vs expression)') +
    theme_minimal_grid() +
    theme(axis.title.y = element_markdown())

# get spearmans corr
growth_cai_df <- meta_df %>% 
  select(Simple_sample_id, `Growth Rate (1/hr)`) %>% 
  inner_join(., exp_cub_corr, by = c("Simple_sample_id" = "condition"))

cor.test(
  x = growth_cai_df$`Growth Rate (1/hr)`, 
  y = growth_cai_df$r2, 
  method = 'spearman'
)


```


```{r, eval=FALSE}
# tAI version

if (metric == "tAI") {
  # get df
  corr_df <- full_df %>% 
    select(locus_tag, tAI) %>% 
    left_join(., exp_df, by = c("locus_tag" = "log-TPM")) 
  # correlations
  exp_cub_corr <- cor(corr_df$tAI, corr_df[3:105], method = "pearson") %>% #3:105 for full
    as_tibble() %>% 
    gather(
      .,
      key = "condition",
      value = "pearsons"
    ) %>% 
    mutate(r2 = pearsons^2)
  
  min_corr <- exp_cub_corr %>% 
    filter(r2 == min(r2))
  max_corr <- exp_cub_corr %>% 
    filter(r2 == max(r2))
  
  fig_corr_min <- corr_df %>% 
    ggplot() +
      aes(x = `cra_crp__delcra_ac`, y = tAI) +    
      geom_point(alpha = 0.5) +
      scale_x_continuous(name = "Condition expression (log TPM)") +
      theme_minimal_grid() +
      labs(title = glue('*R*<sup>2</sup> = {round(min_corr$r2, 3)}')) +
      theme(plot.title = element_markdown())
  
  fig_corr_max <- corr_df %>% 
    ggplot() +
      aes(x = `glu__glu_ale5`, y = tAI) +
      geom_point(alpha = 0.5) +
      scale_x_continuous(name = "Condition expression (log TPM)") +
      theme_minimal_grid() +
      labs(title = glue('*R*<sup>2</sup> = {round(max_corr$r2, 3)}')) +
      theme(plot.title = element_markdown())
  
  fig_corr_grow <- meta_df %>% 
    select(Simple_sample_id, `Growth Rate (1/hr)`) %>% 
    inner_join(., exp_cub_corr, by = c("Simple_sample_id" = "condition")) %>% 
    ggplot() +
      aes(x = `Growth Rate (1/hr)`, y = r2) +
      geom_point(size = 2, alpha = 0.75) +
      scale_y_continuous(name = '*R*<sup>2</sup> (tAI vs expression)') +
      theme_minimal_grid() +
      theme(axis.title.y = element_markdown())
  
} 


# ROC-SEMPPR version

if (metric == "roc_semppr_mean") {
  # get df
  corr_df <- full_df %>% 
    select(locus_tag, roc_semppr_mean) %>% 
    drop_na(roc_semppr_mean) %>% 
    left_join(., exp_df, by = c("locus_tag" = "log-TPM")) 
  # correlations
  exp_cub_corr <- cor(corr_df$roc_semppr_mean, corr_df[3:105], method = "pearson") %>% 
    as_tibble() %>% 
    gather(
      .,
      key = "condition",
      value = "pearsons"
    ) %>% 
    mutate(r2 = pearsons^2)
  
  min_corr <- exp_cub_corr %>% 
    filter(r2 == min(r2))
  max_corr <- exp_cub_corr %>% 
    filter(r2 == max(r2))
  
  fig_corr_min <- corr_df %>% 
    ggplot() +
      aes(x = `fps__fps_serB_ale4`, y = roc_semppr_mean) +    
      geom_point(alpha = 0.5) +
      scale_x_continuous(name = "Condition expression (log TPM)") +
      scale_y_continuous(name = "ROC-SEMPPR") +
      theme_minimal_grid() +
      labs(title = glue('*R*<sup>2</sup> = {round(min_corr$r2, 3)}')) +
      theme(plot.title = element_markdown())
  
  fig_corr_max <- corr_df %>% 
    ggplot() +
      aes(x = `glu__glu_ale5`, y = roc_semppr_mean) +
      geom_point(alpha = 0.5) +
      scale_x_continuous(name = "Condition expression (log TPM)") +
      scale_y_continuous(name = "ROC-SEMPPR") +
      theme_minimal_grid() +
      labs(title = glue('*R*<sup>2</sup> = {round(max_corr$r2, 3)}')) +
      theme(plot.title = element_markdown())
  
  fig_corr_grow <- meta_df %>% 
    select(Simple_sample_id, `Growth Rate (1/hr)`) %>% 
    inner_join(., exp_cub_corr, by = c("Simple_sample_id" = "condition")) %>% 
    ggplot() +
      aes(x = `Growth Rate (1/hr)`, y = r2) +
      geom_point(size = 2, alpha = 0.75) +
      scale_y_continuous(name = '*R*<sup>2</sup> (metric vs expression)') +
      theme_minimal_grid() +
      theme(axis.title.y = element_markdown())
  
} 

```



## Growth rate data

```{r, message=FALSE}

fig_gr_hist <- meta_df %>% 
  ggplot(aes(x = `Growth Rate (1/hr)`)) +
  geom_histogram(binwidth = 0.05) +
  scale_x_continuous(
    limits = c(0, 1.5),
    expand = c(0,0)
  ) +
  scale_y_continuous(
    name = "Count", 
    expand = c(0,0)
  ) +
  theme_minimal_hgrid()

```


## Figure 3
```{r, message=FALSE}

fig_3 <- plot_grid(
  fig_corr_min, fig_corr_max, fig_gr_hist, fig_corr_grow,
  nrow = 2,
  labels = "AUTO"
) +
  theme(plot.background = element_rect(fill="white", color = NA))

fig_3

```

```{r, include=FALSE, eval=FALSE}

save_plot(
  here(
    'Manuscript', 
    'Figures',
    glue('fig_{metric}_exp_gr_{data_tag}.png')   # file name
  ),
  fig_3, 
  base_height = 6.71,
  base_asp = 1.618, 
  base_width = NULL
)

```

```{r, message=FALSE, include=FALSE}

rm(fig_corr_max, fig_corr_min, fig_corr_grow, fig_gr_hist,
   min_corr, max_corr)

```


## Regression analysis to assess predictors of CUB

```{r}
cor(full_df$mean, full_df$lin_r, method = 'pearson')
0.370575*0.370575
cor(full_df$CAI, full_df$lin_r, method = 'pearson')
0.4313216*0.4313216
```


```{r, message=FALSE}

# model with average expression
# use scale() for z score - NOT CURRENTLY USING
lm_exp <- lm(CAI ~ mean, data = full_df)
glance(lm_exp)
tidy(lm_exp)

```


```{r, message=FALSE}

# model with growth correlation
lm_gci <- lm(CAI ~ lin_r, data = full_df)
glance(lm_gci)
tidy(lm_gci)

```


```{r, message=FALSE}

# model with average expression and gci
lm_exp_gci <- lm(CAI ~ mean + lin_r, data = full_df)
glance(lm_exp_gci)
tidy(lm_exp_gci)

# check for collinearity issues
car::vif(lm_exp_gci) 

```


```{r, message=FALSE}

# model with average expression and gci with interaction
lm_inter <- lm(CAI ~ mean + lin_r + mean*lin_r, data = full_df)
glance(lm_inter)
tidy(lm_inter)

```


```{r, message=FALSE}

fig_5a <- tibble(
  model = c(
    "expression",
    "gci",
    "both_add",
    "both_inter"
  ),
  adj_r2 = c(
    glance(lm_exp)$adj.r.squared,
    glance(lm_gci)$adj.r.squared,
    glance(lm_exp_gci)$adj.r.squared,
    glance(lm_inter)$adj.r.squared
  )
) %>% 
  ggplot(
    aes(
      adj_r2,
      fct_relevel(model, "both_inter", "both_add", "gci", "expression")
    )
  ) +
  geom_col() +
  scale_x_continuous(
    name = "Adjusted *R*<sup>2</sup>",
    limits = c(0, .4)
  ) +
  scale_y_discrete(
    labels = c("Both (interaction)", 
               "Both", "GCI", "Expression")
  ) +
  theme_minimal(base_size = 14) +
  theme(
    axis.title.y = element_blank(),
    axis.title.x = element_markdown()
  )

fig_5a

```


```{r, message=FALSE}

# create model fits for range of expression and GCI values for use in plotting
new_data <- expand_grid(
  mean = seq(
    from = max(full_df$mean), 
    to = min(full_df$mean), 
    length.out = 100
  ),
  lin_r = seq(
    from = 1, 
    to = -1, 
    length.out = 7)
 )

gci_models <- new_data %>% 
  mutate(
    pred_cai_lm_exp = predict(lm_exp, new_data),
    pred_cai_lm_both = predict(lm_exp_gci, new_data),
    pred_cai_lm_inter = predict(lm_inter, new_data),
    lin_r = factor(round(lin_r, 3))
  )

# function to create plot 5b expression model
plot_exp_lm <- function(r2_x, r2_y, x_lab) {
  
  # create label data set
  r2_exp_text <- data.frame(
    label = glue(
    '*R*<sup>2</sup> = {round(glance(lm_exp)$adj.r.squared, 3)}'),
    x = r2_x, 
    y = r2_y
  )
  
  # plot expression model
  tibble(
    CAI = full_df$CAI,
    express = full_df$mean
  ) %>% 
    ggplot() +
    geom_point(aes(express, CAI), alpha = 0.1) +
    geom_line(
      data = gci_models, 
      aes(mean, pred_cai_lm_exp), 
      color = "#CC4678FF", # use color of GCI=0 in other plots
      size = 1
    ) +
    # geom_richtext(
    #   data = r2_exp_text,
    #   aes(x = x, y = y,label = label),
    #   label.color = NA,
    #   inherit.aes = FALSE
    # ) +
    scale_x_continuous(name = x_lab) +
    scale_color_viridis_d(option = "C") +
    theme_half_open() +
    background_grid()
  
} 

# function to create plot 5c model with expression & gci (no interaction)
plot_both_lm <- function(r2_x, r2_y, x_lab) {
  
  # create label data set
  r2_both_text <- data.frame(
    label = glue(
    '*R*<sup>2</sup> = {round(glance(lm_exp_gci)$adj.r.squared, 3)}'),
    x = r2_x, 
    y = r2_y
  )
  
  # plot additive model
  tibble(
    CAI = full_df$CAI,
    express = full_df$mean
  ) %>% 
    ggplot() +
    geom_point(aes(express, CAI), alpha = 0.1) +
    geom_line(
      data = gci_models, 
      aes(mean, pred_cai_lm_both, color = fct_rev(lin_r)), 
      size = 1
    ) +
    # geom_richtext(
    #   data = r2_both_text,
    #   aes(x = x, y = y,label = label),
    #   label.color = NA,
    #   inherit.aes = FALSE
    # ) +
    scale_x_continuous(name = x_lab) +
    scale_color_viridis_d(
      option = "C",
      name = "GCI"
    ) +
    theme_half_open() +
    background_grid() +
    theme(legend.position = "none")

}


# function to create plot 5d model with expression & gci with interaction
plot_inter_lm <- function(r2_x, r2_y, x_lab) {
  
  # create label data set
  r2_inter_text <- data.frame(
    label = glue(
    '*R*<sup>2</sup> = {round(glance(lm_inter)$adj.r.squared, 3)}'),
    x = r2_x, 
    y = r2_y
  )
  
  # plot additive model
  tibble(
    CAI = full_df$CAI,
    express = full_df$mean
  ) %>% 
    ggplot() +
    geom_point(aes(express, CAI), alpha = 0.1) +
    geom_line(
      data = gci_models, 
      aes(mean, pred_cai_lm_inter, color = fct_rev(lin_r)), 
      size = 1
    ) +
    # geom_richtext(
    #   data = r2_inter_text,
    #   aes(x = x, y = y,label = label),
    #   label.color = NA,
    #   inherit.aes = FALSE
    # ) +
    scale_x_continuous(name = x_lab) +
    scale_color_viridis_d(
      option = "C",
      name = "GCI"
    ) +
    theme_half_open() +
    background_grid() +
    theme(legend.position = "none")

}

# set values specific to data sets for plots
if (data_tag %in% c("ecoli_full", "ecoli_sparse", "ecoli_neutral", "ecoli_neutral_wt", "ecoli_non_at_rich")) {
  x <- 2
  y <- .8
  lab <- "Expression (mean, log TPM)"
} else {
  x <- 2
  y <- .8
  lab <- "Expression (mean)"
}

# create plots
fig_5b <- plot_exp_lm(x, y, lab)

fig_5c <- plot_both_lm(x, y, lab)

fig_5d <- plot_inter_lm(x, y, lab)

```

## Figure 5
```{r, message=FALSE}

# extract shared legend
legend <- get_legend(
  fig_5c + 
    guides(color = guide_legend(nrow = 1)) +
    theme(legend.position = "bottom")
)

plots <- plot_grid(
  fig_5a, fig_5b, fig_5c, fig_5d,
  nrow = 2,
  labels = "AUTO"
) 

fig_5 <- plot_grid(
  plots, legend,
  nrow = 2,
  rel_heights = c(1, .05)
) +
  theme(plot.background = element_rect(fill="white", color = NA))

fig_5

```

```{r, include=FALSE, eval=FALSE}

save_plot(
  here(
    'Manuscript', 
    'Figures',
    glue('fig_{metric}_gci_model_{data_tag}.png')   # file name
  ),
  fig_5, 
  ncol = 1, 
  nrow = 1, 
  base_height = 6.71, #base_height = 5.71,
  base_asp = 1.618, 
  base_width = NULL
)

```


# Supplemental Figure 1
```{r, include=FALSE, eval=FALSE}

fig_5_sup <- full_df %>% 
  ggplot(aes(x = lin_r, y = CAI)) +
  geom_point(alpha = 0.1) +
  geom_smooth(
    method = "lm", 
    se = FALSE,
    color = "#CC4678FF"
  ) +
  scale_x_continuous(name = "GCI") +
  theme_half_open() +
  background_grid() +
  theme(plot.background = element_rect(fill="white", color = NA))

save_plot(
  here(
    'Manuscript', 
    'Figures',
    glue('fig_{metric}_gci_linear.png')   # file name
  ),
  fig_5_sup, 
  ncol = 1, 
  nrow = 1, 
  base_height = 4.71, #base_height = 5.71,
  base_asp = 1.618, 
  base_width = NULL
)

```


# Supplemental Figure AT-rich genes
```{r}

t_test_at_results <- t.test(
  filter(full_df, at_rich==TRUE)$lin_r, 
  filter(full_df, at_rich==FALSE)$lin_r)
t_test_at_results$p.value  # 2.944941e-64

ttest_text_at <- data.frame(
  label = glue::glue(
    'p-value = {scales::scientific(
    t_test_at_results$p.value, digits = 3
    )}'),
  x = 0.4, 
  y = 1.75
)

fig_at_gci <- full_df %>%  
  mutate(at_rich = case_when(
    at_rich == TRUE ~ 'AT-rich',
    at_rich == FALSE ~ 'not AT-rich')
  ) %>% 
  ggplot(aes(x = lin_r, fill = at_rich)) +
  geom_density(alpha = 0.5) +
  geom_richtext(
    data = ttest_text_at,
    aes(x = x, y = y, label = label),
    label.color = NA,
    inherit.aes = FALSE
  ) +
  scale_x_continuous(name = "GCI", limits = c(-0.9, 0.7)) +
  scale_y_continuous(name = "Density") +
  scale_fill_viridis_d(name = "Gene type", option = "E") +
  theme_minimal_grid() + 
  theme(plot.background = element_rect(fill="white", color = NA))

fig_at_gci

save_plot(
  here('Manuscript', 'Figures', 'fig_at_rich_gci.png'),
  fig_at_gci, 
  ncol = 1, 
  nrow = 1, 
  base_height = 5.71, #base_height = 5.71,
  base_asp = 1.618, 
  base_width = NULL
)

```



# Supplemental tAI and ROC-SEMPPR analysis
```{r, eval=FALSE, message=FALSE, include=FALSE}

if (metric == "tAI" & data_tag == "ecoli_full") {
  
  # fit models
  lm_exp2 <- lm(tAI ~ mean, data = full_df)
  lm_gci2 <- lm(tAI ~ lin_r, data = full_df)
  lm_exp_gci2 <- lm(tAI ~ mean + lin_r, data = full_df)
  lm_inter2 <- lm(tAI ~ mean + lin_r + mean*lin_r, data = full_df)
  
  fig_t_a <- tibble(
    model = c(
      "expression",
      "gci",
      "both_add",
      "both_inter"
    ),
    adj_r2 = c(
      glance(lm_exp2)$adj.r.squared,
      glance(lm_gci2)$adj.r.squared,
      glance(lm_exp_gci2)$adj.r.squared,
      glance(lm_inter2)$adj.r.squared
    )
  ) %>% 
    ggplot(
      aes(
        adj_r2,
        fct_relevel(model, "both_inter", "both_add", "gci", "expression")
      )
    ) +
    geom_col() +
    scale_x_continuous(
      name = "Adjusted *R*<sup>2</sup>",
      limits = c(0, .4)
    ) +
    scale_y_discrete(
      labels = c("Both (interaction)", 
               "Both", "GCI", "Expression")
    ) +
    theme_minimal(base_size = 14) +
    theme(
      axis.title.y = element_blank(),
      axis.title.x = element_markdown()
    )
  
  # create new data sets with model predictions
  new_data2 <- expand_grid(
    mean = seq(
      from = max(full_df$mean), 
      to = min(full_df$mean), 
      length.out = 100
    ),
    lin_r = seq(
      from = 1, 
      to = -1, 
      length.out = 7)
  )

  gci_models2 <- new_data2 %>% 
    mutate(
      pred_tai_lm_exp = predict(lm_exp2, new_data2),
      pred_tai_lm_both = predict(lm_exp_gci2, new_data2),
      pred_tai_lm_inter = predict(lm_inter2, new_data2),
      lin_r = factor(round(lin_r, 3))
  )
  
  # plot models
  # expression model
  fig_t_b <- tibble(
    tAI = full_df$tAI,
    express = full_df$mean
  ) %>% 
    ggplot() +
      geom_point(aes(express, tAI), alpha = 0.1) +
      geom_line(
        data = gci_models2, 
        aes(mean, pred_tai_lm_exp), 
        color = "#CC4678FF", # use color of GCI=0 in other plots
        size = 1
      ) +
      scale_x_continuous(name = "Expression (mean, log TPM)") +
      scale_color_viridis_d(option = "C") +
      theme_half_open() +
      background_grid()
  
  # additive model
  fig_t_c <- tibble(
    tAI = full_df$tAI,
    express = full_df$mean
  ) %>% 
    ggplot() +
    geom_point(aes(express, tAI), alpha = 0.1) +
    geom_line(
      data = gci_models2, 
      aes(mean, pred_tai_lm_both, color = fct_rev(lin_r)), 
      size = 1
    ) +
    scale_x_continuous(name = "Expression (mean, log TPM)") +
    scale_color_viridis_d(
      option = "C",
      name = "GCI"
    ) +
    theme_half_open() +
    background_grid() +
    theme(legend.position = "none")
  
  # interaction model
  fig_t_d <- tibble(
    tAI = full_df$tAI,
    express = full_df$mean
  ) %>% 
    ggplot() +
    geom_point(aes(express, tAI), alpha = 0.1) +
    geom_line(
      data = gci_models2, 
      aes(mean, pred_tai_lm_inter, color = fct_rev(lin_r)), 
      size = 1
    ) +
    scale_x_continuous(name = "Expression (mean, log TPM)") +
    scale_color_viridis_d(
      option = "C",
      name = "GCI"
    ) +
    theme_half_open() +
    background_grid() +
    theme(legend.position = "none")
    
}


```

```{r, eval=FALSE, message=FALSE, include=FALSE}

if (metric == "roc_semppr_mean" & data_tag == "ecoli_full") {
  
  # fit models
  lm_exp2 <- lm(roc_semppr_mean ~ mean, data = full_df)
  lm_gci2 <- lm(roc_semppr_mean ~ lin_r, data = full_df)
  lm_exp_gci2 <- lm(roc_semppr_mean ~ mean + lin_r, data = full_df)
  lm_inter2 <- lm(roc_semppr_mean ~ mean + lin_r + mean*lin_r, data = full_df)
  
  fig_t_a <- tibble(
    model = c(
      "expression",
      "gci",
      "both_add",
      "both_inter"
    ),
    adj_r2 = c(
      glance(lm_exp2)$adj.r.squared,
      glance(lm_gci2)$adj.r.squared,
      glance(lm_exp_gci2)$adj.r.squared,
      glance(lm_inter2)$adj.r.squared
    )
  ) %>% 
    ggplot(
      aes(
        adj_r2,
        fct_relevel(model, "both_inter", "both_add", "gci", "expression")
      )
    ) +
    geom_col() +
    scale_x_continuous(
      name = "Adjusted *R*<sup>2</sup>",
      limits = c(0, .35)
    ) +
    scale_y_discrete(
      labels = c("Both (interaction)", 
               "Both", "GCI", "Expression")
    ) +
    theme_minimal(base_size = 14) +
    theme(
      axis.title.y = element_blank(),
      axis.title.x = element_markdown()
    )
  
  # create new data sets with model predictions
  new_data2 <- expand_grid(
    mean = seq(
      from = max(full_df$mean), 
      to = min(full_df$mean), 
      length.out = 100
    ),
    lin_r = seq(
      from = 1, 
      to = -1, 
      length.out = 7)
  )

  gci_models2 <- new_data2 %>% 
    mutate(
      pred_roc_lm_exp = predict(lm_exp2, new_data2),
      pred_roc_lm_both = predict(lm_exp_gci2, new_data2),
      pred_roc_lm_inter = predict(lm_inter2, new_data2),
      lin_r = factor(round(lin_r, 3))
  )
  
  # plot models
  # expression model
  fig_t_b <- tibble(
    roc_semppr_mean = full_df$roc_semppr_mean,
    express = full_df$mean
  ) %>% 
    ggplot() +
      geom_point(aes(express, roc_semppr_mean), alpha = 0.1) +
      geom_line(
        data = gci_models2, 
        aes(mean, pred_roc_lm_exp), 
        color = "#CC4678FF", # use color of GCI=0 in other plots
        size = 1
      ) +
      scale_x_continuous(name = "Expression (mean, log TPM)") +
      scale_y_continuous(name = "ROC-SEMPPR") +
      scale_color_viridis_d(option = "C") +
      theme_half_open() +
      background_grid()
  
  # additive model
  fig_t_c <- tibble(
    roc_semppr_mean = full_df$roc_semppr_mean,
    express = full_df$mean
  ) %>% 
    ggplot() +
    geom_point(aes(express, roc_semppr_mean), alpha = 0.1) +
    geom_line(
      data = gci_models2, 
      aes(mean, pred_roc_lm_both, color = fct_rev(lin_r)), 
      size = 1
    ) +
    scale_x_continuous(name = "Expression (mean, log TPM)") +
    scale_y_continuous(name = "ROC-SEMPPR") +
    scale_color_viridis_d(
      option = "C",
      name = "GCI"
    ) +
    theme_half_open() +
    background_grid() +
    theme(legend.position = "none")
  
  # interaction model
  fig_t_d <- tibble(
    roc_semppr_mean = full_df$roc_semppr_mean,
    express = full_df$mean
  ) %>% 
    ggplot() +
    geom_point(aes(express, roc_semppr_mean), alpha = 0.1) +
    geom_line(
      data = gci_models2, 
      aes(mean, pred_roc_lm_inter, color = fct_rev(lin_r)), 
      size = 1
    ) +
    scale_x_continuous(name = "Expression (mean, log TPM)") +
    scale_y_continuous(name = "ROC-SEMPPR") +
    scale_color_viridis_d(
      option = "C",
      name = "GCI"
    ) +
    theme_half_open() +
    background_grid() +
    theme(legend.position = "none")
    
}
```



```{r, message=FALSE, eval=FALSE}

# extract shared legend
legend <- get_legend(
  fig_t_c + 
    guides(color = guide_legend(nrow = 1)) +
    theme(legend.position = "bottom")
)

plots2 <- plot_grid(
  fig_t_a, fig_t_b, fig_t_c, fig_t_d,
  nrow = 2,
  labels = "AUTO"
) 

fig_5_tai <- plot_grid(
  plots2, legend,
  nrow = 2,
  rel_heights = c(1, .05)
) +
  theme(plot.background = element_rect(fill="white", color = NA))

fig_5_tai

fig_5_roc <- plot_grid(
  plots2, legend,
  nrow = 2,
  rel_heights = c(1, .05)
) +
  theme(plot.background = element_rect(fill="white", color = NA))

fig_5_roc

```

```{r, include=FALSE, eval=FALSE}

save_plot(
  here(
    'Manuscript', 
    'Figures',
    glue('fig_{metric}_gci_model_{data_tag}.png')   # file name
  ),
  fig_5_roc, 
  ncol = 1, 
  nrow = 1, 
  base_height = 6.71, #base_height = 5.71,
  base_asp = 1.618, 
  base_width = NULL
)

```

### Session Info
```{r}
sessionInfo()
```





