---
title: "Enrichment Gene Analysis"
author: "Luiz Carlos Vieira and Mackenzie Johnson"
output: html_document
---

Functional enrichment analysis is a technique used to interpret and investigate gene lists, derived from 
high-throughput sequence studies.

The clusterProfiler library is designed to perform "over-representation analysis" (ORA), using GO and KEGG
for multiple model organisms and compare functional profiles of various conditions at one level (e.g.,
different treatment groups).

One of the main limitations of ORA is that it restricts the analysis to DE genes, excluding genes that do not
satisfy the chosen significance threshold (usually the vast majority).

[Mais informações sobre clusterProfiler](https://www.sciencedirect.com/science/article/pii/S2666675821000667)

Enrichment of pathways with the terms GO and the KEGG encyclopedia are the most used for functional analysis,
due to its long-standing curation and availability for a wide range of species.

# load libraries
```{r, message=FALSE}
########### Gene Anotation ###########
library(AnnotationDbi)
library(org.EcK12.eg.db)

##### Pathway analysis of DE genes #####
library(clusterProfiler)

# analysis and visualization
library(tidyverse)
library(here)
library(readxl)
library(patchwork)
library(ggtext)
library(cowplot)
```


## Load data
```{r}
data <- read_csv(here('Data', 'genes_GCI', 'ecoli_full_pos_corr.csv'))
neg_data <- read_csv(here('Data', 'genes_GCI', 'ecoli_full_neg_corr.csv'))

head(data)
```


## Mapping gene names to entrzID
```{r}
# check keytypes available
#keytypes(org.EcK12.eg.db)
#columns(org.EcK12.eg.db)

#AnnotationDbi::select(org.EcK12.eg.db, keys=id, columns=c("SYMBOL","GENENAME"), keytype="ENTREZID")

data$ENTREZID <- mapIds(org.EcK12.eg.db,
                              keys = data$gene_name,
                              column = "ENTREZID",
                              keytype = "SYMBOL",
                              multiVals = "first")

neg_data$ENTREZID <- mapIds(org.EcK12.eg.db,
                              keys = neg_data$gene_name,
                              column = "ENTREZID",
                              keytype = "SYMBOL",
                              multiVals = "first")

```


## Removing duplicates gene names
```{r}
data <- dplyr::distinct(data, gene_name, .keep_all= TRUE)

neg_data <- dplyr::distinct(neg_data, gene_name, .keep_all= TRUE)

```


## Creating a list with differential expressed genes

The function enrich...() (gene = a vector with entrez gene_id)
```{r}  
# removing genes without entrezID
mtx <- subset(data, is.na(ENTREZID) == FALSE)
n_mtx <- subset(neg_data, is.na(ENTREZID) == FALSE)

# creating a new matrix
gene_matrix <- mtx$GCI
n_gene_matrix <- n_mtx$GCI

# add entrezID as rownames 
names(gene_matrix) <- mtx$ENTREZID
names(n_gene_matrix) <- n_mtx$ENTREZID

# subseting GCI > 0
# in this case nothing is done
geneList <- gene_matrix[abs(gene_matrix) > 0] 
n_geneList <- n_gene_matrix[abs(n_gene_matrix) > 0]  

head(geneList, 10)
```



# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #
#                                   GO Terms Enrichment
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #

# Gene Ontology [GO](http://geneontology.org/docs/ontology-documentation/)

Gene Ontology defines concepts/classes used to describe gene function and the relationships between these
concepts. It classifies functions in three aspects:

* MF: Molecular Function, molecular activities of gene products

* CC: Cellular Component, where gene products are active

* BP: Biological Process, pathways and major processes constituted by the activities of various gene products

The GO terms are arranged on a directed acyclic graph, where the border between the terms represents the
parental relationship.


## Enrichment GO ALL (MF, CC, BP)
```{r}
go_enrich_ALL <- enrichGO(
  gene=names(geneList),
  OrgDb= 'org.EcK12.eg.db',
  keyType = "ENTREZID",
  ont = "ALL",
  pvalueCutoff = 0.04,
  pAdjustMethod = "BH",
  qvalueCutoff = 0.05,
  readable = TRUE,
  pool = FALSE
)

summary_go_ALL <- as.data.frame(go_enrich_ALL)

go_enrich_ALL_neg <- enrichGO(
  gene=names(n_geneList),
  OrgDb= 'org.EcK12.eg.db',
  keyType = "ENTREZID",
  ont = "ALL",
  pvalueCutoff = 0.04,
  pAdjustMethod = "BH",
  qvalueCutoff = 0.05,
  readable = TRUE,
  pool = FALSE
)

summary_go_ALL_neg <- as.data.frame(go_enrich_ALL_neg)
```

As a result of the enrichment analysis, a table with the following values is generated:

BgRatio, M/N.

M = size of the geneset (eg number of genes on the term).

N = size of all of the unique genes in the collection of genesets


GeneRatio, k/n.

k = only unique genes, the number of genes within that list n (geneList), which are annotated to the term.

n = only unique genes; is the size of the list of genes of interest in the term represented by a gene set


## Calculate Odds Ratio from clusterProfiler GO output
```{r}

df_go_pos_gci <- summary_go_ALL %>% 
  select(c(Description, GeneRatio, BgRatio, p.adjust)) %>% # keep only necessary columns
  extract( # split ratios into numerator and denominator to build contingency table
    GeneRatio, 
    into = c("gene_num", "gene_denom"), 
    regex = "([[:alnum:]]+)/([[:alnum:]]+)"
  ) %>% 
  extract(
    BgRatio,
    into = c("bg_num", "bg_denom"),
    regex = "([[:alnum:]]+)/([[:alnum:]]+)"
  ) %>% 
  dplyr::mutate(
    cat_a = as.numeric(gene_num),
    cat_b = as.numeric(gene_denom) - cat_a,
    cat_c = as.numeric(bg_num) - cat_a,
    cat_d = as.numeric(bg_denom) - cat_a - cat_b - cat_c,
    odds_ratio = (cat_a/cat_b)/(cat_c/cat_d)
  ) 
  
df_go_neg_gci <- summary_go_ALL_neg %>% 
  select(c(Description, GeneRatio, BgRatio, p.adjust)) %>% # keep only necessary columns
  extract( # split ratios into numerator and denominator to build contingency table
    GeneRatio, 
    into = c("gene_num", "gene_denom"), 
    regex = "([[:alnum:]]+)/([[:alnum:]]+)"
  ) %>% 
  extract(
    BgRatio,
    into = c("bg_num", "bg_denom"),
    regex = "([[:alnum:]]+)/([[:alnum:]]+)"
  ) %>% 
  dplyr::mutate(
    cat_a = as.numeric(gene_num),
    cat_b = as.numeric(gene_denom) - cat_a,
    cat_c = as.numeric(bg_num) - cat_a,
    cat_d = as.numeric(bg_denom) - cat_a - cat_b - cat_c,
    odds_ratio = (cat_a/cat_b)/(cat_c/cat_d)
  ) 

```


```{r}

fig_go_pos <- df_go_pos_gci %>% 
  arrange(desc(odds_ratio)) %>% 
  slice(1:15) %>% 
  mutate(labels = str_wrap(Description, 30)) %>% 
  ggplot() +
  aes(x = odds_ratio, y = fct_reorder(labels, odds_ratio)) +
  geom_point(size = 3) +
  geom_vline(xintercept = 1) +
  #scale_x_continuous(name = "Odds Ratio") +
  scale_x_log10(name = "Odds Ratio", limits = c(1, 30)) +
  scale_y_discrete(name = NULL) +
  ggtitle("GO enrichment (GCI > 0)") +
  theme_bw() 

```


```{r}

fig_go_neg <- df_go_neg_gci %>% 
  arrange(desc(odds_ratio)) %>% 
  slice(1:15) %>% 
  mutate(
    labels = str_wrap(Description, 30)
  ) %>% 
  ggplot() +
  geom_point(
    aes(
      x = odds_ratio, 
      y = fct_reorder(labels, odds_ratio)
    ),
    size = 3
  ) +
  geom_vline(xintercept = 1) +
  #scale_x_continuous(name = "Odds Ratio") +
  scale_x_log10(name = "Odds Ratio", limits = c(1, 30)) +
  scale_y_discrete(name = NULL) +
  ggtitle("GO enrichment (GCI < 0)") +
  theme_bw() 

```

```{r}
rm(data, neg_data, mtx, n_mtx, go_enrich_ALL, go_enrich_ALL_neg)
```



## Add in essential vs non-essential / core vs accessory analysis
```{r}

meta_df <- read_tsv(here('Data', 'processed_data', 'processed_metadata_ecoli.tsv'))

exp_df <- read_tsv(here('Data', 'processed_data', 'processed_expression_ecoli.tsv'))

full_df <- read_tsv(here('Data', 'ecoli_info', 'current_ecoli_master_table.tsv')) %>% 
  select(locus_tag, gene)

# essential gene data set with unknown Adam origins
core_df1 <- read_tsv(
  file = here('Data', 'ecoli_info', 'essential_ecoli_genes.txt')
) %>% 
  select(`Gene Name`, `Note`)

# essential gene data set from supp table 7
core_df2 <- read_excel(
  here( 'Data', 'ecoli_info', 'keio_supp_table_7.xls'),
  range = cell_rows(3:4013),
  col_names = c("keio_results", "eck_num", "gene_name", "jw_id", "locus_tag", "cog_id", "cog_cat")
) %>% 
  filter(keio_results == "E") %>% # keep only essential genes
  select(keio_results, gene_name, locus_tag)

# actual core analysis
core_df4 <- 
  read_csv(
    here('Data', 'ecoli_info', 'gene_presence_absence.csv')
  ) %>% 
  filter(`No. isolates` == 60) %>% # 2,449
  filter(`Annotation` != "hypothetical protein") %>%  # 2, 197
  select(`Gene`)

# add gene classifications into data set
full_df <- full_df %>% 
  mutate(
    core1 = case_when(
      gene %in% core_df1$`Gene Name` ~ "essential",
      TRUE ~ "non-essential"
    ),
    core2 = case_when(
      gene %in% core_df2$gene_name ~ "essential",
      TRUE ~ "non-essential"
    ),
    core4 = case_when(
      gene %in% core_df4$`Gene` ~ "core",
      TRUE ~ "accessory"
    )
  ) 

rm(core_df1, core_df2, core_df4)

exp_gr_df <- exp_df %>% 
    gather(-`log-TPM`,
           key = "condition",
           value = "expression") %>% 
    spread(`log-TPM`, expression)

exp_gr_df <- meta_df %>% 
    select(`Simple_sample_id`, `Growth Rate (1/hr)`) %>% 
    inner_join(., exp_gr_df, by = c("Simple_sample_id" = "condition"))

exp_gr_corr <- cor(
  exp_gr_df[2], 
  exp_gr_df[3:ncol(exp_gr_df)], 
  method = "pearson"
) %>% 
  as_tibble() %>% 
  gather(
    .,
    key = "genes",
    value = "pearsons"
  ) %>%
  mutate(r2 = pearsons^2)

exp_gr_corr$genes <- gsub("_[0-9]", "", exp_gr_corr$genes)

# join all data for gene classification and gci values
gci_genomes_df <- exp_gr_corr %>% 
  left_join(., full_df, by = c("genes" = "locus_tag"))

# data set 1
core_gci1 <- gci_genomes_df %>% 
  filter(core1 == "essential") %>% 
  select(pearsons)

access_gci1 <- gci_genomes_df %>% 
  filter(core1 == "non-essential") %>% 
  select(pearsons)

t_test_results <- t.test(core_gci1, access_gci1)
t_test_results$p.value  # 4.434505e-44

# data set 2
core_gci2 <- gci_genomes_df %>% 
  filter(core2 == "essential") %>% 
  select(pearsons)

access_gci2 <- gci_genomes_df %>% 
  filter(core2 == "non-essential") %>% 
  select(pearsons)

t_test_results2 <- t.test(core_gci2, access_gci2)
t_test_results2$p.value  # 2.488719e-36

# data set 4
core_gci4 <- gci_genomes_df %>% 
  filter(core4 == "core") %>% 
  select(pearsons)

access_gci4 <- gci_genomes_df %>% 
  filter(core4 == "accessory") %>% 
  select(pearsons)

t_test_results4 <- t.test(core_gci4, access_gci4)
t_test_results4$p.value  # 3.252126e-65

# use essential data set from Adam in figure
# # create label data set
# ttest_text <- data.frame(
#   label = glue::glue(
#     'p-value = {scales::scientific(
#     t_test_results$p.value, digits = 3
#     )}'),
#   x = 0.3, 
#   y = 1.4
# )

fig_core_gci <- gci_genomes_df %>% 
  drop_na(core1) %>% 
  ggplot(aes(x = pearsons, fill = core1)) +
  geom_density(alpha = 0.5) +
  # geom_richtext(
  #   data = ttest_text,
  #   aes(x = x, y = y, label = label),
  #   label.color = NA,
  #   inherit.aes = FALSE
  # ) +
  scale_x_continuous(name = "GCI") +
  scale_y_continuous(name = "Density") +
  scale_fill_viridis_d(name = "Gene type") +
  theme_minimal_grid()




```


## Figure for core and GO analysis
```{r}

fig_go_plots <- plot_grid(
  fig_go_neg, fig_go_pos,
  nrow = 1
)

fig_6 <- plot_grid(
  fig_core_gci, fig_go_plots,
  nrow = 2,
  rel_heights = c(.5, 1),
  labels = "AUTO"
) + 
  theme(plot.background = element_rect(fill="white", color = NA))

fig_6

```


```{r, include=FALSE, eval=FALSE}

save_plot(
  here('Manuscript', 'Figures', 'fig_core_go_gci.png'),
  fig_6, 
  ncol = 1, 
  nrow = 1, 
  base_height = 6.71, #base_height = 5.71,
  base_asp = 1.618, 
  base_width = NULL
)

```


## Referências:

[Paper clusterProfile](https://www.sciencedirect.com/science/article/pii/S2666675821000667)  
[clusterProfile book](https://yulab-smu.top/biomedical-knowledge-mining-book/index.html)  
[rdocumentation-enrichgo](https://www.rdocumentation.org/packages/clusterProfiler/versions/3.0.4/topics/enrichGO)  
[rdocumentation-enrichkegg](https://www.rdocumentation.org/packages/clusterProfiler/versions/3.0.4/topics/enrichKEGG)  
[rdocumentation-enrichomics](https://www.bioconductor.org/help/course-materials/2017/BioC2017/Day1/Workshops/OmicsData/doc/enrichOmics.html)  



## NOT CURRENTLY RUNNING
### Dotplot

```{r, eval=FALSE}
fig_go_dot <- dotplot(go_enrich_ALL, 
        showCategory= 10, #40,
        title = "GO term enrichment (positive GCI)",
        label_format = 30,
        font.size= 8) 

fig_go_dot

# summary_go_ALL %>% 
#   arrange(desc(Count)) %>% 
#   slice(1:10) %>% 
#   ggplot() + 
#   geom_point(aes(x = GeneRatio, y = Description))

#ggsave(file.path("dotplot_GO_ALL.jpeg"))
```

```{r, eval=FALSE}

fig_go_dot_neg <- dotplot(go_enrich_ALL_neg, 
        showCategory= 10, #40,
        title = "GO term enrichment (negative GCI)",
        label_format = 30,
        font.size= 8) 

fig_go_dot_neg

```


# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #
#                            Enrichment with KEGG database
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #

## Kyoto Encyclopedia of Genes and Genomes (KEGG)

KEGG is a collection of hand-drawn road maps that represent interaction and reaction networks
molecular. These pathways cover a wide range of biochemical processes that can be divided into 7 major
categories:

* Metabolism
* Processing of genetic information
* Environmental information processing
* Cellular processes
* Organizational systems
* human diseases
* Drug development.

To consult the list of organisms supported by the database, see the link below:
[Lista de organismo](https://www.genome.jp/kegg/catalog/org_list.html)


## Enrichment with KEGG
```{r, eval=FALSE}
kegg_enrich <- enrichKEGG(gene= names(geneList),
                          organism = "eco",     
                          keyType = 'ncbi-geneid',
                          pvalueCutoff = 0.04,
                          pAdjustMethod = "BH",
                          qvalueCutoff = 0.05,
                          use_internal_data = FALSE
                          )

# Set gene_id to a human readble format
kegg_enrich.H <- setReadable(kegg_enrich, OrgDb= org.EcK12.eg.db, keyType= "ENTREZID")

# Summary of results:
summary_kegg_enrich <- as.data.frame(kegg_enrich.H)

head(summary_kegg_enrich, 10)

## salving results:
#write.xlsx(summary_kegg_enrich, 'summary_kegg_enrich.xlsx')
```


## KEGG barplot
```{r, eval=FALSE}
fig_kegg_bar <- barplot(kegg_enrich, 
        drop = TRUE, 
        showCategory = 10, #50, 
        title = "KEGG Enrichment Pathways",
        font.size = 8)

fig_kegg_bar

#ggsave(file.path("barplot_Kegg.jpeg"))
```


## Kegg Dotplot
```{r, eval=FALSE}
fig_kegg_dot <- dotplot(kegg_enrich, 
        showCategory = 10, #40,
        title = "Kegg Enrichment",
        label_format = 30,
        font.size= 9)

fig_kegg_dot

#ggsave(file.path("dotplot_Kegg.jpeg"))
```


### Session Info
```{r}
sessionInfo()
```
